# Using Word2Vec for Topic Modeling

The idea of using an $N$ dimensional feature vector for unsupervised learning a powerful concept.  If we can take a body of words and consider them as a vector, we can begin to interpret words and sentences as points in $N$ dimensional space, and use traditional methods, including the K Means algorithm, to assign labels to unstructured data.

In fact, the most complex algorithm in this notebook is *not* the K Means algorithm, but rather the preparation of the words to be used by the algorithm, which is the Word2Vec algorithm.  Word2Vec is a member of the class of algorithms that carry out a *word embedding*, which simply means that we take a collection of words and return a feature vector.  The feature vector is always of the same dimension, even though the collection of words may not be.   

An amazing property of word2vec is that vector embeddings can often result in interesting geometric interpretations, including additive properties like 

$$\text{w2v}(king) - \text{w2v}(queen) = \text{w2v}(man) - \text{w2v}(woman)$$, 

which suggests that an article, or body of text, can be a superposition of word vectors whose position in feature space relative to other articles can indicate their similarity.  This is a very helpful insight.

Let's take a look at this algorithm in action.  First, let's import some libraries:

In [1]:
// command line operations
import scala.sys.process._

// Spark Classes
import scala.collection.mutable
//import org.apache.spark.mllib.clustering.LDA
//import org.apache.spark.mllib.linalg.{Vector, Vectors}
import org.apache.spark.rdd.RDD
import org.apache.spark.ml.clustering.KMeans
import org.apache.spark.ml.feature.Word2Vec
import org.apache.spark.mllib.linalg.{Vector, Vectors}

// .tar and .gz utilities
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.commons.compress.compressors.gzip.GzipCompressorInputStream
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.apache.spark.input.PortableDataStream
import scala.util.Try
import java.nio.charset._

// rounding function for time:
def round(x: Long) ={Math.round(x/1e6)/1e3}

# Loading the Corpus

We will download the yahoo 20 newsgroups dataset if you haven't already done so. This dataset contains 20,000 news stories that have been manually categorized into directories.  

In [2]:
import sys.process._
val listing = "ls 20_newsgroups"!! ;

if (listing.contains("20_newsgroups") == false ) {
    println("downloading dataset")
    "wget http://qwone.com/~jason/20Newsgroups/20news-19997.tar.gz"!;
    "tar -xvf 20news-19997.tar.gz"!;
    "rm 20news-19997*"!;
}
println("directories in 20 newsgroup dataset")


--2018-11-25 20:15:31--  http://qwone.com/~jason/20Newsgroups/20news-19997.tar.gz
downloading dataset
Resolving proxy1.fyre.ibm.com (proxy1.fyre.ibm.com)... 172.16.200.60, 172.16.200.61
Connecting to proxy1.fyre.ibm.com (proxy1.fyre.ibm.com)|172.16.200.60|:3128... connected.
Proxy request sent, awaiting response... 200 OK
Length: 17332201 (17M) [application/x-gzip]
Saving to: ‘20news-19997.tar.gz.4’

     0K .......... .......... .......... .......... ..........  0% 58.5M 0s
    50K .......... .......... .......... .......... ..........  0% 80.5M 0s
   100K .......... .......... .......... .......... ..........  0% 80.0M 0s
   150K .......... .......... .......... .......... ..........  1%  103M 0s
   200K .......... .......... .......... .......... ..........  1% 93.6M 0s
   250K .......... .......... .......... .......... ..........  1%  112M 0s
   300K .......... .......... .......... .......... ..........  2% 86.3M 0s
   350K .......... .......... .......... .......... ..........  

  5200K .......... .......... .......... .......... .......... 31%  870K 1s
  5250K .......... .......... .......... .......... .......... 31%  116M 0s
  5300K .......... .......... .......... .......... .......... 31% 86.2M 0s
  5350K .......... .......... .......... .......... .......... 31% 79.1M 0s
  5400K .......... .......... .......... .......... .......... 32%  103M 0s
  5450K .......... .......... .......... .......... .......... 32% 81.6M 0s
  5500K .......... .......... .......... .......... .......... 32% 73.1M 0s
  5550K .......... .......... .......... .......... .......... 33% 75.7M 0s
  5600K .......... .......... .......... .......... .......... 33% 84.0M 0s
  5650K .......... .......... .......... .......... .......... 33% 98.6M 0s
  5700K .......... .......... .......... .......... .......... 33% 97.9M 0s
  5750K .......... .......... .......... .......... .......... 34%  109M 0s
  5800K .......... .......... .......... .......... .......... 34%  120M 0s
  5850K ....

 11550K .......... .......... .......... .......... .......... 68% 95.7M 0s
 11600K .......... .......... .......... .......... .......... 68%  124M 0s
 11650K .......... .......... .......... .......... .......... 69%  126M 0s
 11700K .......... .......... .......... .......... .......... 69%  140M 0s
 11750K .......... .......... .......... .......... .......... 69%  115M 0s
 11800K .......... .......... .......... .......... .......... 70%  137M 0s
 11850K .......... .......... .......... .......... .......... 70%  104M 0s
 11900K .......... .......... .......... .......... .......... 70% 86.0M 0s
 11950K .......... .......... .......... .......... .......... 70% 86.2M 0s
 12000K .......... .......... .......... .......... .......... 71% 95.3M 0s
 12050K .......... .......... .......... .......... .......... 71%  109M 0s
 12100K .......... .......... .......... .......... .......... 71%  110M 0s
 12150K .......... .......... .......... .......... .......... 72%  116M 0s
 12200K ....

20_newsgroups/
20_newsgroups/alt.atheism/
20_newsgroups/alt.atheism/53366
20_newsgroups/alt.atheism/53367
20_newsgroups/alt.atheism/51247
20_newsgroups/alt.atheism/51248
20_newsgroups/alt.atheism/51249
20_newsgroups/alt.atheism/51250
20_newsgroups/alt.atheism/51251
20_newsgroups/alt.atheism/51252
20_newsgroups/alt.atheism/51253
20_newsgroups/alt.atheism/51254
20_newsgroups/alt.atheism/51255
20_newsgroups/alt.atheism/53070
20_newsgroups/alt.atheism/53071
20_newsgroups/alt.atheism/53072
20_newsgroups/alt.atheism/53090
20_newsgroups/alt.atheism/53134
20_newsgroups/alt.atheism/53135
20_newsgroups/alt.atheism/53136
20_newsgroups/alt.atheism/53137
20_newsgroups/alt.atheism/53138
20_newsgroups/alt.atheism/53139
20_newsgroups/alt.atheism/53140
20_newsgroups/alt.atheism/53141
20_newsgroups/alt.atheism/53142
20_newsgroups/alt.atheism/54125
20_newsgroups/alt.atheism/53213
20_newsgroups/alt.atheism/53214
20_newsgroups/alt.atheism/53215
20_newsgroups/alt.atheism/53216
20_newsgroups/alt.atheism/5321

20_newsgroups/alt.atheism/53584
20_newsgroups/alt.atheism/53585
20_newsgroups/alt.atheism/53586
20_newsgroups/alt.atheism/53587
20_newsgroups/alt.atheism/53588
20_newsgroups/alt.atheism/53589
20_newsgroups/alt.atheism/53590
20_newsgroups/alt.atheism/53591
20_newsgroups/alt.atheism/53592
20_newsgroups/alt.atheism/53593
20_newsgroups/alt.atheism/53594
20_newsgroups/alt.atheism/51167
20_newsgroups/alt.atheism/53595
20_newsgroups/alt.atheism/53596
20_newsgroups/alt.atheism/53597
20_newsgroups/alt.atheism/53598
20_newsgroups/alt.atheism/53599
20_newsgroups/alt.atheism/53600
20_newsgroups/alt.atheism/53601
20_newsgroups/alt.atheism/53602
20_newsgroups/alt.atheism/53603
20_newsgroups/alt.atheism/53604
20_newsgroups/alt.atheism/53605
20_newsgroups/alt.atheism/53606
20_newsgroups/alt.atheism/51275
20_newsgroups/alt.atheism/53607
20_newsgroups/alt.atheism/53609
20_newsgroups/alt.atheism/53610
20_newsgroups/alt.atheism/51276
20_newsgroups/alt.atheism/51119
20_newsgroups/alt.atheism/51120
20_newsg

20_newsgroups/alt.atheism/53152
20_newsgroups/alt.atheism/53153
20_newsgroups/alt.atheism/53154
20_newsgroups/alt.atheism/53155
20_newsgroups/alt.atheism/53156
20_newsgroups/alt.atheism/53157
20_newsgroups/alt.atheism/53158
20_newsgroups/alt.atheism/53159
20_newsgroups/alt.atheism/53160
20_newsgroups/alt.atheism/53161
20_newsgroups/alt.atheism/53162
20_newsgroups/alt.atheism/53163
20_newsgroups/alt.atheism/53164
20_newsgroups/alt.atheism/53165
20_newsgroups/alt.atheism/53166
20_newsgroups/alt.atheism/53167
20_newsgroups/alt.atheism/53168
20_newsgroups/alt.atheism/53169
20_newsgroups/alt.atheism/53170
20_newsgroups/alt.atheism/53171
20_newsgroups/alt.atheism/53172
20_newsgroups/alt.atheism/53173
20_newsgroups/alt.atheism/53174
20_newsgroups/alt.atheism/53175
20_newsgroups/alt.atheism/53176
20_newsgroups/alt.atheism/53177
20_newsgroups/alt.atheism/53178
20_newsgroups/alt.atheism/53179
20_newsgroups/alt.atheism/53180
20_newsgroups/alt.atheism/53181
20_newsgroups/alt.atheism/53182
20_newsg

20_newsgroups/alt.atheism/54196
20_newsgroups/alt.atheism/54197
20_newsgroups/alt.atheism/54198
20_newsgroups/alt.atheism/54199
20_newsgroups/alt.atheism/54200
20_newsgroups/alt.atheism/54201
20_newsgroups/alt.atheism/54202
20_newsgroups/alt.atheism/54203
20_newsgroups/alt.atheism/54204
20_newsgroups/alt.atheism/54215
20_newsgroups/alt.atheism/54216
20_newsgroups/alt.atheism/54217
20_newsgroups/alt.atheism/54218
20_newsgroups/alt.atheism/54219
20_newsgroups/alt.atheism/54220
20_newsgroups/alt.atheism/54221
20_newsgroups/alt.atheism/54222
20_newsgroups/alt.atheism/54223
20_newsgroups/alt.atheism/54224
20_newsgroups/alt.atheism/54225
20_newsgroups/alt.atheism/54226
20_newsgroups/alt.atheism/54227
20_newsgroups/alt.atheism/54228
20_newsgroups/alt.atheism/54229
20_newsgroups/alt.atheism/54230
20_newsgroups/alt.atheism/54232
20_newsgroups/alt.atheism/54233
20_newsgroups/alt.atheism/54234
20_newsgroups/alt.atheism/54235
20_newsgroups/alt.atheism/54236
20_newsgroups/alt.atheism/54237
20_newsg

20_newsgroups/comp.graphics/38238
20_newsgroups/comp.graphics/38239
20_newsgroups/comp.graphics/38240
20_newsgroups/comp.graphics/38241
20_newsgroups/comp.graphics/38242
20_newsgroups/comp.graphics/38243
20_newsgroups/comp.graphics/38244
20_newsgroups/comp.graphics/38245
20_newsgroups/comp.graphics/38246
20_newsgroups/comp.graphics/38247
20_newsgroups/comp.graphics/38248
20_newsgroups/comp.graphics/38249
20_newsgroups/comp.graphics/38250
20_newsgroups/comp.graphics/38251
20_newsgroups/comp.graphics/38252
20_newsgroups/comp.graphics/38253
20_newsgroups/comp.graphics/38254
20_newsgroups/comp.graphics/38255
20_newsgroups/comp.graphics/38256
20_newsgroups/comp.graphics/38257
20_newsgroups/comp.graphics/38258
20_newsgroups/comp.graphics/38259
20_newsgroups/comp.graphics/38260
20_newsgroups/comp.graphics/37261
20_newsgroups/comp.graphics/38261
20_newsgroups/comp.graphics/38262
20_newsgroups/comp.graphics/38263
20_newsgroups/comp.graphics/38264
20_newsgroups/comp.graphics/38265
20_newsgroups/

20_newsgroups/comp.graphics/38606
20_newsgroups/comp.graphics/38607
20_newsgroups/comp.graphics/38608
20_newsgroups/comp.graphics/38609
20_newsgroups/comp.graphics/38610
20_newsgroups/comp.graphics/38694
20_newsgroups/comp.graphics/38611
20_newsgroups/comp.graphics/38657
20_newsgroups/comp.graphics/38658
20_newsgroups/comp.graphics/38659
20_newsgroups/comp.graphics/38660
20_newsgroups/comp.graphics/38661
20_newsgroups/comp.graphics/38662
20_newsgroups/comp.graphics/38705
20_newsgroups/comp.graphics/38706
20_newsgroups/comp.graphics/38707
20_newsgroups/comp.graphics/38708
20_newsgroups/comp.graphics/38663
20_newsgroups/comp.graphics/38752
20_newsgroups/comp.graphics/38665
20_newsgroups/comp.graphics/38666
20_newsgroups/comp.graphics/38613
20_newsgroups/comp.graphics/38614
20_newsgroups/comp.graphics/38615
20_newsgroups/comp.graphics/38616
20_newsgroups/comp.graphics/38617
20_newsgroups/comp.graphics/38618
20_newsgroups/comp.graphics/38619
20_newsgroups/comp.graphics/38620
20_newsgroups/

20_newsgroups/comp.graphics/38848
20_newsgroups/comp.graphics/38849
20_newsgroups/comp.graphics/38850
20_newsgroups/comp.graphics/38852
20_newsgroups/comp.graphics/38853
20_newsgroups/comp.graphics/38854
20_newsgroups/comp.graphics/38855
20_newsgroups/comp.graphics/38856
20_newsgroups/comp.graphics/38857
20_newsgroups/comp.graphics/38862
20_newsgroups/comp.graphics/38863
20_newsgroups/comp.graphics/38864
20_newsgroups/comp.graphics/38865
20_newsgroups/comp.graphics/38866
20_newsgroups/comp.graphics/38867
20_newsgroups/comp.graphics/38868
20_newsgroups/comp.graphics/38869
20_newsgroups/comp.graphics/38913
20_newsgroups/comp.graphics/38870
20_newsgroups/comp.graphics/38876
20_newsgroups/comp.graphics/38877
20_newsgroups/comp.graphics/38878
20_newsgroups/comp.graphics/38879
20_newsgroups/comp.graphics/38880
20_newsgroups/comp.graphics/38881
20_newsgroups/comp.graphics/38914
20_newsgroups/comp.graphics/38915
20_newsgroups/comp.graphics/38918
20_newsgroups/comp.graphics/38926
20_newsgroups/

20_newsgroups/comp.graphics/39642
20_newsgroups/comp.graphics/39643
20_newsgroups/comp.graphics/39644
20_newsgroups/comp.graphics/39645
20_newsgroups/comp.graphics/39646
20_newsgroups/comp.graphics/39647
20_newsgroups/comp.graphics/39648
20_newsgroups/comp.graphics/39649
20_newsgroups/comp.graphics/39650
20_newsgroups/comp.graphics/39651
20_newsgroups/comp.graphics/39652
20_newsgroups/comp.graphics/39653
20_newsgroups/comp.graphics/39654
20_newsgroups/comp.graphics/39655
20_newsgroups/comp.graphics/39656
20_newsgroups/comp.graphics/39657
20_newsgroups/comp.graphics/39658
20_newsgroups/comp.graphics/39659
20_newsgroups/comp.graphics/39660
20_newsgroups/comp.graphics/39661
20_newsgroups/comp.graphics/39662
20_newsgroups/comp.graphics/39663
20_newsgroups/comp.graphics/39500
20_newsgroups/comp.graphics/39664
20_newsgroups/comp.graphics/39665
20_newsgroups/comp.graphics/39666
20_newsgroups/comp.graphics/39667
20_newsgroups/comp.graphics/39668
20_newsgroups/comp.graphics/39669
20_newsgroups/

20_newsgroups/comp.os.ms-windows.misc/9584
20_newsgroups/comp.os.ms-windows.misc/9585
20_newsgroups/comp.os.ms-windows.misc/9586
20_newsgroups/comp.os.ms-windows.misc/9587
20_newsgroups/comp.os.ms-windows.misc/9588
20_newsgroups/comp.os.ms-windows.misc/9589
20_newsgroups/comp.os.ms-windows.misc/9590
20_newsgroups/comp.os.ms-windows.misc/9591
20_newsgroups/comp.os.ms-windows.misc/9592
20_newsgroups/comp.os.ms-windows.misc/9593
20_newsgroups/comp.os.ms-windows.misc/9594
20_newsgroups/comp.os.ms-windows.misc/9595
20_newsgroups/comp.os.ms-windows.misc/9596
20_newsgroups/comp.os.ms-windows.misc/9597
20_newsgroups/comp.os.ms-windows.misc/9598
20_newsgroups/comp.os.ms-windows.misc/9599
20_newsgroups/comp.os.ms-windows.misc/9600
20_newsgroups/comp.os.ms-windows.misc/9601
20_newsgroups/comp.os.ms-windows.misc/9602
20_newsgroups/comp.os.ms-windows.misc/9603
20_newsgroups/comp.os.ms-windows.misc/9604
20_newsgroups/comp.os.ms-windows.misc/9605
20_newsgroups/comp.os.ms-windows.misc/9606
20_newsgrou

20_newsgroups/comp.os.ms-windows.misc/9862
20_newsgroups/comp.os.ms-windows.misc/9863
20_newsgroups/comp.os.ms-windows.misc/9864
20_newsgroups/comp.os.ms-windows.misc/9865
20_newsgroups/comp.os.ms-windows.misc/9866
20_newsgroups/comp.os.ms-windows.misc/9867
20_newsgroups/comp.os.ms-windows.misc/9868
20_newsgroups/comp.os.ms-windows.misc/9869
20_newsgroups/comp.os.ms-windows.misc/9870
20_newsgroups/comp.os.ms-windows.misc/9871
20_newsgroups/comp.os.ms-windows.misc/9872
20_newsgroups/comp.os.ms-windows.misc/9873
20_newsgroups/comp.os.ms-windows.misc/9874
20_newsgroups/comp.os.ms-windows.misc/9875
20_newsgroups/comp.os.ms-windows.misc/9876
20_newsgroups/comp.os.ms-windows.misc/9877
20_newsgroups/comp.os.ms-windows.misc/9878
20_newsgroups/comp.os.ms-windows.misc/9879
20_newsgroups/comp.os.ms-windows.misc/9880
20_newsgroups/comp.os.ms-windows.misc/9881
20_newsgroups/comp.os.ms-windows.misc/9882
20_newsgroups/comp.os.ms-windows.misc/9883
20_newsgroups/comp.os.ms-windows.misc/9884
20_newsgrou

20_newsgroups/comp.os.ms-windows.misc/10016
20_newsgroups/comp.os.ms-windows.misc/10017
20_newsgroups/comp.os.ms-windows.misc/10018
20_newsgroups/comp.os.ms-windows.misc/10019
20_newsgroups/comp.os.ms-windows.misc/10020
20_newsgroups/comp.os.ms-windows.misc/10021
20_newsgroups/comp.os.ms-windows.misc/10022
20_newsgroups/comp.os.ms-windows.misc/10023
20_newsgroups/comp.os.ms-windows.misc/10024
20_newsgroups/comp.os.ms-windows.misc/10025
20_newsgroups/comp.os.ms-windows.misc/10026
20_newsgroups/comp.os.ms-windows.misc/10027
20_newsgroups/comp.os.ms-windows.misc/10028
20_newsgroups/comp.os.ms-windows.misc/10029
20_newsgroups/comp.os.ms-windows.misc/10052
20_newsgroups/comp.os.ms-windows.misc/10053
20_newsgroups/comp.os.ms-windows.misc/10054
20_newsgroups/comp.os.ms-windows.misc/10055
20_newsgroups/comp.os.ms-windows.misc/10056
20_newsgroups/comp.os.ms-windows.misc/10057
20_newsgroups/comp.os.ms-windows.misc/10058
20_newsgroups/comp.os.ms-windows.misc/10059
20_newsgroups/comp.os.ms-windows

20_newsgroups/comp.os.ms-windows.misc/10645
20_newsgroups/comp.os.ms-windows.misc/10646
20_newsgroups/comp.os.ms-windows.misc/10853
20_newsgroups/comp.os.ms-windows.misc/10647
20_newsgroups/comp.os.ms-windows.misc/10648
20_newsgroups/comp.os.ms-windows.misc/10649
20_newsgroups/comp.os.ms-windows.misc/10672
20_newsgroups/comp.os.ms-windows.misc/10673
20_newsgroups/comp.os.ms-windows.misc/10674
20_newsgroups/comp.os.ms-windows.misc/10675
20_newsgroups/comp.os.ms-windows.misc/10683
20_newsgroups/comp.os.ms-windows.misc/10684
20_newsgroups/comp.os.ms-windows.misc/10685
20_newsgroups/comp.os.ms-windows.misc/10692
20_newsgroups/comp.os.ms-windows.misc/10693
20_newsgroups/comp.os.ms-windows.misc/10694
20_newsgroups/comp.os.ms-windows.misc/10695
20_newsgroups/comp.os.ms-windows.misc/10696
20_newsgroups/comp.os.ms-windows.misc/10697
20_newsgroups/comp.os.ms-windows.misc/10698
20_newsgroups/comp.os.ms-windows.misc/10723
20_newsgroups/comp.os.ms-windows.misc/10724
20_newsgroups/comp.os.ms-windows

20_newsgroups/comp.sys.ibm.pc.hardware/60203
20_newsgroups/comp.sys.ibm.pc.hardware/60204
20_newsgroups/comp.sys.ibm.pc.hardware/60205
20_newsgroups/comp.sys.ibm.pc.hardware/60206
20_newsgroups/comp.sys.ibm.pc.hardware/60207
20_newsgroups/comp.sys.ibm.pc.hardware/60208
20_newsgroups/comp.sys.ibm.pc.hardware/60209
20_newsgroups/comp.sys.ibm.pc.hardware/60210
20_newsgroups/comp.sys.ibm.pc.hardware/60211
20_newsgroups/comp.sys.ibm.pc.hardware/60212
20_newsgroups/comp.sys.ibm.pc.hardware/60213
20_newsgroups/comp.sys.ibm.pc.hardware/60214
20_newsgroups/comp.sys.ibm.pc.hardware/60215
20_newsgroups/comp.sys.ibm.pc.hardware/60216
20_newsgroups/comp.sys.ibm.pc.hardware/60217
20_newsgroups/comp.sys.ibm.pc.hardware/60218
20_newsgroups/comp.sys.ibm.pc.hardware/60219
20_newsgroups/comp.sys.ibm.pc.hardware/60220
20_newsgroups/comp.sys.ibm.pc.hardware/60221
20_newsgroups/comp.sys.ibm.pc.hardware/60222
20_newsgroups/comp.sys.ibm.pc.hardware/60223
20_newsgroups/comp.sys.ibm.pc.hardware/60224
20_newsgro

20_newsgroups/comp.sys.ibm.pc.hardware/60755
20_newsgroups/comp.sys.ibm.pc.hardware/60756
20_newsgroups/comp.sys.ibm.pc.hardware/60757
20_newsgroups/comp.sys.ibm.pc.hardware/60758
20_newsgroups/comp.sys.ibm.pc.hardware/60759
20_newsgroups/comp.sys.ibm.pc.hardware/60760
20_newsgroups/comp.sys.ibm.pc.hardware/60761
20_newsgroups/comp.sys.ibm.pc.hardware/60762
20_newsgroups/comp.sys.ibm.pc.hardware/60763
20_newsgroups/comp.sys.ibm.pc.hardware/60764
20_newsgroups/comp.sys.ibm.pc.hardware/60765
20_newsgroups/comp.sys.ibm.pc.hardware/60766
20_newsgroups/comp.sys.ibm.pc.hardware/60767
20_newsgroups/comp.sys.ibm.pc.hardware/60768
20_newsgroups/comp.sys.ibm.pc.hardware/60769
20_newsgroups/comp.sys.ibm.pc.hardware/60770
20_newsgroups/comp.sys.ibm.pc.hardware/60771
20_newsgroups/comp.sys.ibm.pc.hardware/60772
20_newsgroups/comp.sys.ibm.pc.hardware/60773
20_newsgroups/comp.sys.ibm.pc.hardware/60774
20_newsgroups/comp.sys.ibm.pc.hardware/60775
20_newsgroups/comp.sys.ibm.pc.hardware/60776
20_newsgro

20_newsgroups/comp.sys.ibm.pc.hardware/61027
20_newsgroups/comp.sys.ibm.pc.hardware/61028
20_newsgroups/comp.sys.ibm.pc.hardware/61029
20_newsgroups/comp.sys.ibm.pc.hardware/61030
20_newsgroups/comp.sys.ibm.pc.hardware/61031
20_newsgroups/comp.sys.ibm.pc.hardware/61032
20_newsgroups/comp.sys.ibm.pc.hardware/61033
20_newsgroups/comp.sys.ibm.pc.hardware/61034
20_newsgroups/comp.sys.ibm.pc.hardware/61035
20_newsgroups/comp.sys.ibm.pc.hardware/61036
20_newsgroups/comp.sys.ibm.pc.hardware/61037
20_newsgroups/comp.sys.ibm.pc.hardware/61038
20_newsgroups/comp.sys.ibm.pc.hardware/61039
20_newsgroups/comp.sys.ibm.pc.hardware/61040
20_newsgroups/comp.sys.ibm.pc.hardware/61041
20_newsgroups/comp.sys.ibm.pc.hardware/61042
20_newsgroups/comp.sys.ibm.pc.hardware/61043
20_newsgroups/comp.sys.ibm.pc.hardware/61044
20_newsgroups/comp.sys.ibm.pc.hardware/61045
20_newsgroups/comp.sys.ibm.pc.hardware/61046
20_newsgroups/comp.sys.ibm.pc.hardware/61047
20_newsgroups/comp.sys.ibm.pc.hardware/61048
20_newsgro

20_newsgroups/comp.sys.ibm.pc.hardware/61009
20_newsgroups/comp.sys.ibm.pc.hardware/61010
20_newsgroups/comp.sys.ibm.pc.hardware/61011
20_newsgroups/comp.sys.ibm.pc.hardware/61012
20_newsgroups/comp.sys.ibm.pc.hardware/61013
20_newsgroups/comp.sys.ibm.pc.hardware/60487
20_newsgroups/comp.sys.ibm.pc.hardware/61014
20_newsgroups/comp.sys.ibm.pc.hardware/61061
20_newsgroups/comp.sys.ibm.pc.hardware/61062
20_newsgroups/comp.sys.ibm.pc.hardware/61063
20_newsgroups/comp.sys.ibm.pc.hardware/61065
20_newsgroups/comp.sys.ibm.pc.hardware/61066
20_newsgroups/comp.sys.ibm.pc.hardware/61067
20_newsgroups/comp.sys.ibm.pc.hardware/61068
20_newsgroups/comp.sys.ibm.pc.hardware/61069
20_newsgroups/comp.sys.ibm.pc.hardware/61070
20_newsgroups/comp.sys.ibm.pc.hardware/61071
20_newsgroups/comp.sys.ibm.pc.hardware/61072
20_newsgroups/comp.sys.ibm.pc.hardware/61073
20_newsgroups/comp.sys.ibm.pc.hardware/61074
20_newsgroups/comp.sys.ibm.pc.hardware/61075
20_newsgroups/comp.sys.ibm.pc.hardware/61076
20_newsgro

20_newsgroups/comp.sys.ibm.pc.hardware/60153
20_newsgroups/comp.sys.ibm.pc.hardware/58860
20_newsgroups/comp.sys.ibm.pc.hardware/60154
20_newsgroups/comp.sys.ibm.pc.hardware/60155
20_newsgroups/comp.sys.ibm.pc.hardware/60156
20_newsgroups/comp.sys.ibm.pc.hardware/60157
20_newsgroups/comp.sys.ibm.pc.hardware/60158
20_newsgroups/comp.sys.ibm.pc.hardware/60159
20_newsgroups/comp.sys.ibm.pc.hardware/60160
20_newsgroups/comp.sys.ibm.pc.hardware/60161
20_newsgroups/comp.sys.ibm.pc.hardware/60162
20_newsgroups/comp.sys.ibm.pc.hardware/60163
20_newsgroups/comp.sys.ibm.pc.hardware/60164
20_newsgroups/comp.sys.ibm.pc.hardware/60165
20_newsgroups/comp.sys.ibm.pc.hardware/60166
20_newsgroups/comp.sys.ibm.pc.hardware/60167
20_newsgroups/comp.sys.ibm.pc.hardware/60168
20_newsgroups/comp.sys.ibm.pc.hardware/60169
20_newsgroups/comp.sys.ibm.pc.hardware/60170
20_newsgroups/comp.sys.ibm.pc.hardware/60171
20_newsgroups/comp.sys.ibm.pc.hardware/60172
20_newsgroups/comp.sys.ibm.pc.hardware/60173
20_newsgro

20_newsgroups/comp.sys.mac.hardware/52012
20_newsgroups/comp.sys.mac.hardware/52013
20_newsgroups/comp.sys.mac.hardware/52014
20_newsgroups/comp.sys.mac.hardware/52015
20_newsgroups/comp.sys.mac.hardware/52016
20_newsgroups/comp.sys.mac.hardware/52017
20_newsgroups/comp.sys.mac.hardware/52018
20_newsgroups/comp.sys.mac.hardware/52019
20_newsgroups/comp.sys.mac.hardware/52020
20_newsgroups/comp.sys.mac.hardware/52021
20_newsgroups/comp.sys.mac.hardware/52022
20_newsgroups/comp.sys.mac.hardware/52023
20_newsgroups/comp.sys.mac.hardware/52024
20_newsgroups/comp.sys.mac.hardware/52025
20_newsgroups/comp.sys.mac.hardware/52026
20_newsgroups/comp.sys.mac.hardware/52027
20_newsgroups/comp.sys.mac.hardware/52028
20_newsgroups/comp.sys.mac.hardware/52029
20_newsgroups/comp.sys.mac.hardware/52030
20_newsgroups/comp.sys.mac.hardware/52031
20_newsgroups/comp.sys.mac.hardware/52032
20_newsgroups/comp.sys.mac.hardware/52033
20_newsgroups/comp.sys.mac.hardware/52034
20_newsgroups/comp.sys.mac.hardwar

20_newsgroups/comp.sys.mac.hardware/51736
20_newsgroups/comp.sys.mac.hardware/51737
20_newsgroups/comp.sys.mac.hardware/51738
20_newsgroups/comp.sys.mac.hardware/51739
20_newsgroups/comp.sys.mac.hardware/51740
20_newsgroups/comp.sys.mac.hardware/51741
20_newsgroups/comp.sys.mac.hardware/51742
20_newsgroups/comp.sys.mac.hardware/51743
20_newsgroups/comp.sys.mac.hardware/51744
20_newsgroups/comp.sys.mac.hardware/51745
20_newsgroups/comp.sys.mac.hardware/51746
20_newsgroups/comp.sys.mac.hardware/51747
20_newsgroups/comp.sys.mac.hardware/51748
20_newsgroups/comp.sys.mac.hardware/51749
20_newsgroups/comp.sys.mac.hardware/51750
20_newsgroups/comp.sys.mac.hardware/51751
20_newsgroups/comp.sys.mac.hardware/52201
20_newsgroups/comp.sys.mac.hardware/51752
20_newsgroups/comp.sys.mac.hardware/51753
20_newsgroups/comp.sys.mac.hardware/51754
20_newsgroups/comp.sys.mac.hardware/51755
20_newsgroups/comp.sys.mac.hardware/51756
20_newsgroups/comp.sys.mac.hardware/51757
20_newsgroups/comp.sys.mac.hardwar

20_newsgroups/comp.sys.mac.hardware/51840
20_newsgroups/comp.sys.mac.hardware/51841
20_newsgroups/comp.sys.mac.hardware/51842
20_newsgroups/comp.sys.mac.hardware/51843
20_newsgroups/comp.sys.mac.hardware/51844
20_newsgroups/comp.sys.mac.hardware/51845
20_newsgroups/comp.sys.mac.hardware/51846
20_newsgroups/comp.sys.mac.hardware/51847
20_newsgroups/comp.sys.mac.hardware/51848
20_newsgroups/comp.sys.mac.hardware/51849
20_newsgroups/comp.sys.mac.hardware/51850
20_newsgroups/comp.sys.mac.hardware/51851
20_newsgroups/comp.sys.mac.hardware/51852
20_newsgroups/comp.sys.mac.hardware/51853
20_newsgroups/comp.sys.mac.hardware/51854
20_newsgroups/comp.sys.mac.hardware/51855
20_newsgroups/comp.sys.mac.hardware/51856
20_newsgroups/comp.sys.mac.hardware/51857
20_newsgroups/comp.sys.mac.hardware/51858
20_newsgroups/comp.sys.mac.hardware/51859
20_newsgroups/comp.sys.mac.hardware/51860
20_newsgroups/comp.sys.mac.hardware/51861
20_newsgroups/comp.sys.mac.hardware/51862
20_newsgroups/comp.sys.mac.hardwar

20_newsgroups/comp.sys.mac.hardware/52307
20_newsgroups/comp.sys.mac.hardware/52308
20_newsgroups/comp.sys.mac.hardware/52309
20_newsgroups/comp.sys.mac.hardware/52310
20_newsgroups/comp.sys.mac.hardware/52311
20_newsgroups/comp.sys.mac.hardware/52312
20_newsgroups/comp.sys.mac.hardware/52313
20_newsgroups/comp.sys.mac.hardware/52314
20_newsgroups/comp.sys.mac.hardware/52315
20_newsgroups/comp.sys.mac.hardware/52316
20_newsgroups/comp.sys.mac.hardware/52317
20_newsgroups/comp.sys.mac.hardware/52318
20_newsgroups/comp.sys.mac.hardware/52319
20_newsgroups/comp.sys.mac.hardware/52320
20_newsgroups/comp.sys.mac.hardware/52321
20_newsgroups/comp.sys.mac.hardware/52322
20_newsgroups/comp.sys.mac.hardware/52323
20_newsgroups/comp.sys.mac.hardware/52324
20_newsgroups/comp.sys.mac.hardware/52325
20_newsgroups/comp.sys.mac.hardware/52326
20_newsgroups/comp.sys.mac.hardware/52327
20_newsgroups/comp.sys.mac.hardware/52328
20_newsgroups/comp.sys.mac.hardware/52329
20_newsgroups/comp.sys.mac.hardwar

20_newsgroups/comp.windows.x/66959
20_newsgroups/comp.windows.x/66960
20_newsgroups/comp.windows.x/66961
20_newsgroups/comp.windows.x/67185
20_newsgroups/comp.windows.x/66962
20_newsgroups/comp.windows.x/66963
20_newsgroups/comp.windows.x/67186
20_newsgroups/comp.windows.x/66964
20_newsgroups/comp.windows.x/67187
20_newsgroups/comp.windows.x/67188
20_newsgroups/comp.windows.x/66965
20_newsgroups/comp.windows.x/66966
20_newsgroups/comp.windows.x/66967
20_newsgroups/comp.windows.x/66968
20_newsgroups/comp.windows.x/66969
20_newsgroups/comp.windows.x/66970
20_newsgroups/comp.windows.x/66971
20_newsgroups/comp.windows.x/66972
20_newsgroups/comp.windows.x/66973
20_newsgroups/comp.windows.x/66974
20_newsgroups/comp.windows.x/66975
20_newsgroups/comp.windows.x/66976
20_newsgroups/comp.windows.x/66977
20_newsgroups/comp.windows.x/66978
20_newsgroups/comp.windows.x/66979
20_newsgroups/comp.windows.x/66980
20_newsgroups/comp.windows.x/66981
20_newsgroups/comp.windows.x/66982
20_newsgroups/comp.w

20_newsgroups/comp.windows.x/66927
20_newsgroups/comp.windows.x/66928
20_newsgroups/comp.windows.x/66929
20_newsgroups/comp.windows.x/66930
20_newsgroups/comp.windows.x/66931
20_newsgroups/comp.windows.x/66932
20_newsgroups/comp.windows.x/66933
20_newsgroups/comp.windows.x/66934
20_newsgroups/comp.windows.x/66935
20_newsgroups/comp.windows.x/66189
20_newsgroups/comp.windows.x/66936
20_newsgroups/comp.windows.x/66937
20_newsgroups/comp.windows.x/66938
20_newsgroups/comp.windows.x/66939
20_newsgroups/comp.windows.x/66940
20_newsgroups/comp.windows.x/66941
20_newsgroups/comp.windows.x/66942
20_newsgroups/comp.windows.x/66943
20_newsgroups/comp.windows.x/66944
20_newsgroups/comp.windows.x/66945
20_newsgroups/comp.windows.x/66946
20_newsgroups/comp.windows.x/66947
20_newsgroups/comp.windows.x/66948
20_newsgroups/comp.windows.x/66949
20_newsgroups/comp.windows.x/66950
20_newsgroups/comp.windows.x/66951
20_newsgroups/comp.windows.x/67197
20_newsgroups/comp.windows.x/67198
20_newsgroups/comp.w

20_newsgroups/comp.windows.x/67534
20_newsgroups/comp.windows.x/67535
20_newsgroups/comp.windows.x/67536
20_newsgroups/comp.windows.x/67537
20_newsgroups/comp.windows.x/67538
20_newsgroups/comp.windows.x/67539
20_newsgroups/comp.windows.x/67540
20_newsgroups/comp.windows.x/67541
20_newsgroups/comp.windows.x/67542
20_newsgroups/comp.windows.x/67543
20_newsgroups/comp.windows.x/67544
20_newsgroups/comp.windows.x/67545
20_newsgroups/comp.windows.x/67546
20_newsgroups/comp.windows.x/67547
20_newsgroups/comp.windows.x/67548
20_newsgroups/comp.windows.x/67549
20_newsgroups/comp.windows.x/67550
20_newsgroups/comp.windows.x/67551
20_newsgroups/comp.windows.x/67553
20_newsgroups/comp.windows.x/67554
20_newsgroups/comp.windows.x/67555
20_newsgroups/comp.windows.x/67556
20_newsgroups/comp.windows.x/67557
20_newsgroups/comp.windows.x/67558
20_newsgroups/comp.windows.x/67559
20_newsgroups/comp.windows.x/67560
20_newsgroups/comp.windows.x/68170
20_newsgroups/comp.windows.x/67972
20_newsgroups/comp.w

20_newsgroups/comp.windows.x/68002
20_newsgroups/comp.windows.x/68003
20_newsgroups/comp.windows.x/68004
20_newsgroups/comp.windows.x/68005
20_newsgroups/comp.windows.x/68006
20_newsgroups/comp.windows.x/68007
20_newsgroups/comp.windows.x/68008
20_newsgroups/comp.windows.x/68009
20_newsgroups/comp.windows.x/68013
20_newsgroups/comp.windows.x/68014
20_newsgroups/comp.windows.x/68019
20_newsgroups/comp.windows.x/68020
20_newsgroups/misc.forsale/
20_newsgroups/misc.forsale/75967
20_newsgroups/misc.forsale/76527
20_newsgroups/misc.forsale/75842
20_newsgroups/misc.forsale/75968
20_newsgroups/misc.forsale/75969
20_newsgroups/misc.forsale/75970
20_newsgroups/misc.forsale/76280
20_newsgroups/misc.forsale/76281
20_newsgroups/misc.forsale/76282
20_newsgroups/misc.forsale/76291
20_newsgroups/misc.forsale/76292
20_newsgroups/misc.forsale/75971
20_newsgroups/misc.forsale/76293
20_newsgroups/misc.forsale/76298
20_newsgroups/misc.forsale/76299
20_newsgroups/misc.forsale/76300
20_newsgroups/misc.forsa

20_newsgroups/misc.forsale/76147
20_newsgroups/misc.forsale/76148
20_newsgroups/misc.forsale/76149
20_newsgroups/misc.forsale/76150
20_newsgroups/misc.forsale/76151
20_newsgroups/misc.forsale/76152
20_newsgroups/misc.forsale/76153
20_newsgroups/misc.forsale/76154
20_newsgroups/misc.forsale/76155
20_newsgroups/misc.forsale/76156
20_newsgroups/misc.forsale/76157
20_newsgroups/misc.forsale/76158
20_newsgroups/misc.forsale/76159
20_newsgroups/misc.forsale/76160
20_newsgroups/misc.forsale/76161
20_newsgroups/misc.forsale/76018
20_newsgroups/misc.forsale/76162
20_newsgroups/misc.forsale/76163
20_newsgroups/misc.forsale/76164
20_newsgroups/misc.forsale/76165
20_newsgroups/misc.forsale/76166
20_newsgroups/misc.forsale/76167
20_newsgroups/misc.forsale/76168
20_newsgroups/misc.forsale/76169
20_newsgroups/misc.forsale/76170
20_newsgroups/misc.forsale/76171
20_newsgroups/misc.forsale/76172
20_newsgroups/misc.forsale/76173
20_newsgroups/misc.forsale/76174
20_newsgroups/misc.forsale/76175
20_newsgro

20_newsgroups/misc.forsale/76445
20_newsgroups/misc.forsale/76446
20_newsgroups/misc.forsale/76447
20_newsgroups/misc.forsale/76448
20_newsgroups/misc.forsale/76854
20_newsgroups/misc.forsale/76449
20_newsgroups/misc.forsale/76450
20_newsgroups/misc.forsale/76451
20_newsgroups/misc.forsale/76452
20_newsgroups/misc.forsale/76453
20_newsgroups/misc.forsale/76454
20_newsgroups/misc.forsale/76455
20_newsgroups/misc.forsale/76456
20_newsgroups/misc.forsale/76457
20_newsgroups/misc.forsale/76458
20_newsgroups/misc.forsale/76459
20_newsgroups/misc.forsale/76460
20_newsgroups/misc.forsale/76461
20_newsgroups/misc.forsale/76462
20_newsgroups/misc.forsale/76463
20_newsgroups/misc.forsale/76464
20_newsgroups/misc.forsale/76465
20_newsgroups/misc.forsale/76466
20_newsgroups/misc.forsale/76467
20_newsgroups/misc.forsale/76468
20_newsgroups/misc.forsale/76469
20_newsgroups/misc.forsale/76470
20_newsgroups/misc.forsale/76471
20_newsgroups/misc.forsale/76472
20_newsgroups/misc.forsale/76473
20_newsgro

20_newsgroups/rec.autos/102893
20_newsgroups/rec.autos/102902
20_newsgroups/rec.autos/102903
20_newsgroups/rec.autos/102778
20_newsgroups/rec.autos/101558
20_newsgroups/rec.autos/102904
20_newsgroups/rec.autos/102779
20_newsgroups/rec.autos/101559
20_newsgroups/rec.autos/101560
20_newsgroups/rec.autos/101561
20_newsgroups/rec.autos/101562
20_newsgroups/rec.autos/101563
20_newsgroups/rec.autos/101627
20_newsgroups/rec.autos/102780
20_newsgroups/rec.autos/102781
20_newsgroups/rec.autos/101628
20_newsgroups/rec.autos/102905
20_newsgroups/rec.autos/101629
20_newsgroups/rec.autos/101630
20_newsgroups/rec.autos/102782
20_newsgroups/rec.autos/101631
20_newsgroups/rec.autos/101632
20_newsgroups/rec.autos/101633
20_newsgroups/rec.autos/101634
20_newsgroups/rec.autos/101635
20_newsgroups/rec.autos/101660
20_newsgroups/rec.autos/102783
20_newsgroups/rec.autos/101661
20_newsgroups/rec.autos/102784
20_newsgroups/rec.autos/102785
20_newsgroups/rec.autos/101662
20_newsgroups/rec.autos/102786
20_newsg

20_newsgroups/rec.autos/103082
20_newsgroups/rec.autos/103084
20_newsgroups/rec.autos/103085
20_newsgroups/rec.autos/103086
20_newsgroups/rec.autos/103146
20_newsgroups/rec.autos/103147
20_newsgroups/rec.autos/103148
20_newsgroups/rec.autos/103149
20_newsgroups/rec.autos/103160
20_newsgroups/rec.autos/103161
20_newsgroups/rec.autos/103168
20_newsgroups/rec.autos/103169
20_newsgroups/rec.autos/103170
20_newsgroups/rec.autos/103171
20_newsgroups/rec.autos/103172
20_newsgroups/rec.autos/103173
20_newsgroups/rec.autos/101551
20_newsgroups/rec.autos/101552
20_newsgroups/rec.autos/101553
20_newsgroups/rec.autos/101554
20_newsgroups/rec.autos/101580
20_newsgroups/rec.autos/101581
20_newsgroups/rec.autos/101582
20_newsgroups/rec.autos/101583
20_newsgroups/rec.autos/101584
20_newsgroups/rec.autos/101585
20_newsgroups/rec.autos/101586
20_newsgroups/rec.autos/101587
20_newsgroups/rec.autos/101588
20_newsgroups/rec.autos/101589
20_newsgroups/rec.autos/101590
20_newsgroups/rec.autos/101591
20_newsg

20_newsgroups/rec.autos/103268
20_newsgroups/rec.autos/103269
20_newsgroups/rec.autos/103270
20_newsgroups/rec.autos/103278
20_newsgroups/rec.autos/103279
20_newsgroups/rec.autos/103647
20_newsgroups/rec.autos/103280
20_newsgroups/rec.autos/103281
20_newsgroups/rec.autos/103282
20_newsgroups/rec.autos/103283
20_newsgroups/rec.autos/103405
20_newsgroups/rec.autos/103284
20_newsgroups/rec.autos/103292
20_newsgroups/rec.autos/103293
20_newsgroups/rec.autos/103294
20_newsgroups/rec.autos/103295
20_newsgroups/rec.autos/103296
20_newsgroups/rec.autos/103297
20_newsgroups/rec.autos/103298
20_newsgroups/rec.autos/103299
20_newsgroups/rec.autos/103300
20_newsgroups/rec.autos/103301
20_newsgroups/rec.autos/103302
20_newsgroups/rec.autos/103303
20_newsgroups/rec.autos/103406
20_newsgroups/rec.autos/103304
20_newsgroups/rec.autos/103306
20_newsgroups/rec.autos/103307
20_newsgroups/rec.autos/103308
20_newsgroups/rec.autos/103309
20_newsgroups/rec.autos/103310
20_newsgroups/rec.autos/103311
20_newsg

20_newsgroups/rec.autos/103754
20_newsgroups/rec.autos/103755
20_newsgroups/rec.autos/103756
20_newsgroups/rec.autos/103757
20_newsgroups/rec.autos/103758
20_newsgroups/rec.autos/103768
20_newsgroups/rec.autos/103760
20_newsgroups/rec.autos/103761
20_newsgroups/rec.autos/103762
20_newsgroups/rec.autos/103763
20_newsgroups/rec.autos/103764
20_newsgroups/rec.autos/103765
20_newsgroups/rec.autos/103766
20_newsgroups/rec.autos/103767
20_newsgroups/rec.autos/103769
20_newsgroups/rec.autos/103774
20_newsgroups/rec.autos/103775
20_newsgroups/rec.autos/103776
20_newsgroups/rec.autos/103777
20_newsgroups/rec.autos/103781
20_newsgroups/rec.motorcycles/
20_newsgroups/rec.motorcycles/104787
20_newsgroups/rec.motorcycles/104388
20_newsgroups/rec.motorcycles/104282
20_newsgroups/rec.motorcycles/104788
20_newsgroups/rec.motorcycles/104389
20_newsgroups/rec.motorcycles/104390
20_newsgroups/rec.motorcycles/104316
20_newsgroups/rec.motorcycles/104789
20_newsgroups/rec.motorcycles/104790
20_newsgroups/re

20_newsgroups/rec.motorcycles/104563
20_newsgroups/rec.motorcycles/104564
20_newsgroups/rec.motorcycles/104565
20_newsgroups/rec.motorcycles/104566
20_newsgroups/rec.motorcycles/104567
20_newsgroups/rec.motorcycles/104568
20_newsgroups/rec.motorcycles/104569
20_newsgroups/rec.motorcycles/103125
20_newsgroups/rec.motorcycles/103126
20_newsgroups/rec.motorcycles/103127
20_newsgroups/rec.motorcycles/103131
20_newsgroups/rec.motorcycles/103132
20_newsgroups/rec.motorcycles/103133
20_newsgroups/rec.motorcycles/103134
20_newsgroups/rec.motorcycles/103137
20_newsgroups/rec.motorcycles/103138
20_newsgroups/rec.motorcycles/103139
20_newsgroups/rec.motorcycles/103140
20_newsgroups/rec.motorcycles/103141
20_newsgroups/rec.motorcycles/103142
20_newsgroups/rec.motorcycles/103143
20_newsgroups/rec.motorcycles/103144
20_newsgroups/rec.motorcycles/103226
20_newsgroups/rec.motorcycles/103227
20_newsgroups/rec.motorcycles/103228
20_newsgroups/rec.motorcycles/103229
20_newsgroups/rec.motorcycles/103230
2

20_newsgroups/rec.motorcycles/104784
20_newsgroups/rec.motorcycles/104785
20_newsgroups/rec.motorcycles/104786
20_newsgroups/rec.motorcycles/104791
20_newsgroups/rec.motorcycles/104792
20_newsgroups/rec.motorcycles/104793
20_newsgroups/rec.motorcycles/104794
20_newsgroups/rec.motorcycles/103553
20_newsgroups/rec.motorcycles/104795
20_newsgroups/rec.motorcycles/104796
20_newsgroups/rec.motorcycles/104797
20_newsgroups/rec.motorcycles/104798
20_newsgroups/rec.motorcycles/104801
20_newsgroups/rec.motorcycles/104802
20_newsgroups/rec.motorcycles/104803
20_newsgroups/rec.motorcycles/104804
20_newsgroups/rec.motorcycles/104805
20_newsgroups/rec.motorcycles/104806
20_newsgroups/rec.motorcycles/104807
20_newsgroups/rec.motorcycles/104808
20_newsgroups/rec.motorcycles/104809
20_newsgroups/rec.motorcycles/104853
20_newsgroups/rec.motorcycles/104854
20_newsgroups/rec.motorcycles/104855
20_newsgroups/rec.motorcycles/104856
20_newsgroups/rec.motorcycles/104857
20_newsgroups/rec.motorcycles/104858
2

20_newsgroups/rec.motorcycles/104947
20_newsgroups/rec.motorcycles/104948
20_newsgroups/rec.motorcycles/104949
20_newsgroups/rec.motorcycles/104950
20_newsgroups/rec.motorcycles/104952
20_newsgroups/rec.motorcycles/104953
20_newsgroups/rec.motorcycles/104954
20_newsgroups/rec.motorcycles/104955
20_newsgroups/rec.motorcycles/104956
20_newsgroups/rec.motorcycles/104957
20_newsgroups/rec.motorcycles/104958
20_newsgroups/rec.motorcycles/104959
20_newsgroups/rec.motorcycles/104960
20_newsgroups/rec.motorcycles/104961
20_newsgroups/rec.motorcycles/104962
20_newsgroups/rec.motorcycles/104963
20_newsgroups/rec.motorcycles/104964
20_newsgroups/rec.motorcycles/104965
20_newsgroups/rec.motorcycles/104966
20_newsgroups/rec.motorcycles/104967
20_newsgroups/rec.motorcycles/105097
20_newsgroups/rec.motorcycles/104968
20_newsgroups/rec.motorcycles/105098
20_newsgroups/rec.motorcycles/105099
20_newsgroups/rec.motorcycles/105100
20_newsgroups/rec.motorcycles/105101
20_newsgroups/rec.motorcycles/104969
2

20_newsgroups/rec.sport.baseball/104350
20_newsgroups/rec.sport.baseball/104351
20_newsgroups/rec.sport.baseball/104352
20_newsgroups/rec.sport.baseball/104353
20_newsgroups/rec.sport.baseball/104354
20_newsgroups/rec.sport.baseball/104355
20_newsgroups/rec.sport.baseball/104356
20_newsgroups/rec.sport.baseball/104357
20_newsgroups/rec.sport.baseball/104358
20_newsgroups/rec.sport.baseball/104359
20_newsgroups/rec.sport.baseball/104360
20_newsgroups/rec.sport.baseball/104507
20_newsgroups/rec.sport.baseball/104361
20_newsgroups/rec.sport.baseball/104362
20_newsgroups/rec.sport.baseball/104363
20_newsgroups/rec.sport.baseball/104508
20_newsgroups/rec.sport.baseball/104509
20_newsgroups/rec.sport.baseball/104510
20_newsgroups/rec.sport.baseball/104511
20_newsgroups/rec.sport.baseball/104665
20_newsgroups/rec.sport.baseball/104688
20_newsgroups/rec.sport.baseball/104689
20_newsgroups/rec.sport.baseball/104690
20_newsgroups/rec.sport.baseball/104691
20_newsgroups/rec.sport.baseball/104692


20_newsgroups/rec.sport.baseball/104516
20_newsgroups/rec.sport.baseball/104517
20_newsgroups/rec.sport.baseball/104518
20_newsgroups/rec.sport.baseball/104519
20_newsgroups/rec.sport.baseball/104520
20_newsgroups/rec.sport.baseball/104521
20_newsgroups/rec.sport.baseball/104522
20_newsgroups/rec.sport.baseball/104523
20_newsgroups/rec.sport.baseball/104524
20_newsgroups/rec.sport.baseball/104525
20_newsgroups/rec.sport.baseball/104526
20_newsgroups/rec.sport.baseball/104527
20_newsgroups/rec.sport.baseball/104528
20_newsgroups/rec.sport.baseball/104529
20_newsgroups/rec.sport.baseball/104530
20_newsgroups/rec.sport.baseball/104531
20_newsgroups/rec.sport.baseball/102151
20_newsgroups/rec.sport.baseball/104532
20_newsgroups/rec.sport.baseball/104533
20_newsgroups/rec.sport.baseball/104534
20_newsgroups/rec.sport.baseball/104535
20_newsgroups/rec.sport.baseball/104536
20_newsgroups/rec.sport.baseball/104537
20_newsgroups/rec.sport.baseball/104538
20_newsgroups/rec.sport.baseball/104539


20_newsgroups/rec.sport.baseball/101666
20_newsgroups/rec.sport.baseball/102654
20_newsgroups/rec.sport.baseball/102655
20_newsgroups/rec.sport.baseball/102656
20_newsgroups/rec.sport.baseball/102657
20_newsgroups/rec.sport.baseball/102658
20_newsgroups/rec.sport.baseball/102667
20_newsgroups/rec.sport.baseball/102668
20_newsgroups/rec.sport.baseball/102669
20_newsgroups/rec.sport.baseball/102670
20_newsgroups/rec.sport.baseball/102671
20_newsgroups/rec.sport.baseball/102672
20_newsgroups/rec.sport.baseball/102673
20_newsgroups/rec.sport.baseball/102674
20_newsgroups/rec.sport.baseball/102675
20_newsgroups/rec.sport.baseball/102676
20_newsgroups/rec.sport.baseball/102677
20_newsgroups/rec.sport.baseball/102678
20_newsgroups/rec.sport.baseball/102679
20_newsgroups/rec.sport.baseball/102680
20_newsgroups/rec.sport.baseball/102681
20_newsgroups/rec.sport.baseball/102682
20_newsgroups/rec.sport.baseball/102683
20_newsgroups/rec.sport.baseball/102684
20_newsgroups/rec.sport.baseball/102685


20_newsgroups/rec.sport.baseball/104940
20_newsgroups/rec.sport.baseball/104941
20_newsgroups/rec.sport.baseball/104942
20_newsgroups/rec.sport.baseball/104943
20_newsgroups/rec.sport.baseball/104944
20_newsgroups/rec.sport.baseball/104945
20_newsgroups/rec.sport.baseball/104946
20_newsgroups/rec.sport.baseball/104947
20_newsgroups/rec.sport.baseball/104948
20_newsgroups/rec.sport.baseball/104949
20_newsgroups/rec.sport.baseball/104950
20_newsgroups/rec.sport.baseball/104951
20_newsgroups/rec.sport.baseball/104952
20_newsgroups/rec.sport.baseball/104953
20_newsgroups/rec.sport.baseball/104954
20_newsgroups/rec.sport.baseball/104955
20_newsgroups/rec.sport.baseball/104956
20_newsgroups/rec.sport.baseball/104957
20_newsgroups/rec.sport.baseball/104958
20_newsgroups/rec.sport.baseball/104959
20_newsgroups/rec.sport.baseball/104960
20_newsgroups/rec.sport.baseball/104961
20_newsgroups/rec.sport.baseball/104962
20_newsgroups/rec.sport.baseball/104963
20_newsgroups/rec.sport.baseball/104964


20_newsgroups/rec.sport.hockey/53607
20_newsgroups/rec.sport.hockey/53608
20_newsgroups/rec.sport.hockey/53609
20_newsgroups/rec.sport.hockey/53610
20_newsgroups/rec.sport.hockey/53611
20_newsgroups/rec.sport.hockey/53612
20_newsgroups/rec.sport.hockey/53613
20_newsgroups/rec.sport.hockey/53614
20_newsgroups/rec.sport.hockey/53615
20_newsgroups/rec.sport.hockey/53616
20_newsgroups/rec.sport.hockey/53617
20_newsgroups/rec.sport.hockey/53618
20_newsgroups/rec.sport.hockey/53619
20_newsgroups/rec.sport.hockey/53620
20_newsgroups/rec.sport.hockey/53621
20_newsgroups/rec.sport.hockey/53622
20_newsgroups/rec.sport.hockey/53623
20_newsgroups/rec.sport.hockey/53624
20_newsgroups/rec.sport.hockey/53625
20_newsgroups/rec.sport.hockey/53626
20_newsgroups/rec.sport.hockey/53627
20_newsgroups/rec.sport.hockey/53628
20_newsgroups/rec.sport.hockey/53629
20_newsgroups/rec.sport.hockey/53630
20_newsgroups/rec.sport.hockey/53631
20_newsgroups/rec.sport.hockey/53632
20_newsgroups/rec.sport.hockey/53633
2

20_newsgroups/rec.sport.hockey/53659
20_newsgroups/rec.sport.hockey/53660
20_newsgroups/rec.sport.hockey/53661
20_newsgroups/rec.sport.hockey/53662
20_newsgroups/rec.sport.hockey/53663
20_newsgroups/rec.sport.hockey/53664
20_newsgroups/rec.sport.hockey/53673
20_newsgroups/rec.sport.hockey/53674
20_newsgroups/rec.sport.hockey/53675
20_newsgroups/rec.sport.hockey/53676
20_newsgroups/rec.sport.hockey/53677
20_newsgroups/rec.sport.hockey/53678
20_newsgroups/rec.sport.hockey/53679
20_newsgroups/rec.sport.hockey/53680
20_newsgroups/rec.sport.hockey/53681
20_newsgroups/rec.sport.hockey/53682
20_newsgroups/rec.sport.hockey/53683
20_newsgroups/rec.sport.hockey/53684
20_newsgroups/rec.sport.hockey/53685
20_newsgroups/rec.sport.hockey/53686
20_newsgroups/rec.sport.hockey/53687
20_newsgroups/rec.sport.hockey/53688
20_newsgroups/rec.sport.hockey/53689
20_newsgroups/rec.sport.hockey/53690
20_newsgroups/rec.sport.hockey/53691
20_newsgroups/rec.sport.hockey/53692
20_newsgroups/rec.sport.hockey/53693
2

20_newsgroups/rec.sport.hockey/54011
20_newsgroups/rec.sport.hockey/54012
20_newsgroups/rec.sport.hockey/54013
20_newsgroups/rec.sport.hockey/54014
20_newsgroups/rec.sport.hockey/54015
20_newsgroups/rec.sport.hockey/54016
20_newsgroups/rec.sport.hockey/54017
20_newsgroups/rec.sport.hockey/54018
20_newsgroups/rec.sport.hockey/54019
20_newsgroups/rec.sport.hockey/54020
20_newsgroups/rec.sport.hockey/54021
20_newsgroups/rec.sport.hockey/54022
20_newsgroups/rec.sport.hockey/54023
20_newsgroups/rec.sport.hockey/54024
20_newsgroups/rec.sport.hockey/54025
20_newsgroups/rec.sport.hockey/54026
20_newsgroups/rec.sport.hockey/54027
20_newsgroups/rec.sport.hockey/54028
20_newsgroups/rec.sport.hockey/54036
20_newsgroups/rec.sport.hockey/54040
20_newsgroups/rec.sport.hockey/54041
20_newsgroups/rec.sport.hockey/54042
20_newsgroups/rec.sport.hockey/54043
20_newsgroups/rec.sport.hockey/54044
20_newsgroups/rec.sport.hockey/54045
20_newsgroups/rec.sport.hockey/54046
20_newsgroups/rec.sport.hockey/54047
2

20_newsgroups/rec.sport.hockey/54706
20_newsgroups/rec.sport.hockey/54707
20_newsgroups/rec.sport.hockey/54708
20_newsgroups/rec.sport.hockey/54709
20_newsgroups/rec.sport.hockey/54710
20_newsgroups/rec.sport.hockey/54711
20_newsgroups/rec.sport.hockey/54712
20_newsgroups/rec.sport.hockey/54714
20_newsgroups/rec.sport.hockey/54715
20_newsgroups/rec.sport.hockey/54716
20_newsgroups/rec.sport.hockey/54717
20_newsgroups/rec.sport.hockey/54718
20_newsgroups/rec.sport.hockey/54719
20_newsgroups/rec.sport.hockey/54720
20_newsgroups/rec.sport.hockey/54721
20_newsgroups/rec.sport.hockey/54722
20_newsgroups/rec.sport.hockey/54723
20_newsgroups/rec.sport.hockey/54724
20_newsgroups/rec.sport.hockey/54725
20_newsgroups/rec.sport.hockey/54726
20_newsgroups/rec.sport.hockey/54727
20_newsgroups/rec.sport.hockey/54728
20_newsgroups/rec.sport.hockey/54729
20_newsgroups/rec.sport.hockey/54730
20_newsgroups/rec.sport.hockey/54731
20_newsgroups/rec.sport.hockey/54732
20_newsgroups/rec.sport.hockey/54733
2

20_newsgroups/sci.crypt/15596
20_newsgroups/sci.crypt/15003
20_newsgroups/sci.crypt/15187
20_newsgroups/sci.crypt/15188
20_newsgroups/sci.crypt/15844
20_newsgroups/sci.crypt/15845
20_newsgroups/sci.crypt/15846
20_newsgroups/sci.crypt/15847
20_newsgroups/sci.crypt/15848
20_newsgroups/sci.crypt/15849
20_newsgroups/sci.crypt/15850
20_newsgroups/sci.crypt/15851
20_newsgroups/sci.crypt/15852
20_newsgroups/sci.crypt/15189
20_newsgroups/sci.crypt/15190
20_newsgroups/sci.crypt/15191
20_newsgroups/sci.crypt/15192
20_newsgroups/sci.crypt/15193
20_newsgroups/sci.crypt/15194
20_newsgroups/sci.crypt/15195
20_newsgroups/sci.crypt/15196
20_newsgroups/sci.crypt/15197
20_newsgroups/sci.crypt/15198
20_newsgroups/sci.crypt/15199
20_newsgroups/sci.crypt/15200
20_newsgroups/sci.crypt/15201
20_newsgroups/sci.crypt/15202
20_newsgroups/sci.crypt/15203
20_newsgroups/sci.crypt/15204
20_newsgroups/sci.crypt/15205
20_newsgroups/sci.crypt/15206
20_newsgroups/sci.crypt/15207
20_newsgroups/sci.crypt/15208
20_newsgro

20_newsgroups/sci.crypt/15543
20_newsgroups/sci.crypt/15544
20_newsgroups/sci.crypt/15545
20_newsgroups/sci.crypt/15546
20_newsgroups/sci.crypt/15547
20_newsgroups/sci.crypt/15548
20_newsgroups/sci.crypt/15550
20_newsgroups/sci.crypt/15551
20_newsgroups/sci.crypt/15552
20_newsgroups/sci.crypt/15553
20_newsgroups/sci.crypt/15554
20_newsgroups/sci.crypt/15555
20_newsgroups/sci.crypt/15556
20_newsgroups/sci.crypt/15557
20_newsgroups/sci.crypt/15597
20_newsgroups/sci.crypt/15599
20_newsgroups/sci.crypt/15600
20_newsgroups/sci.crypt/15601
20_newsgroups/sci.crypt/15602
20_newsgroups/sci.crypt/15605
20_newsgroups/sci.crypt/15606
20_newsgroups/sci.crypt/15608
20_newsgroups/sci.crypt/15609
20_newsgroups/sci.crypt/15611
20_newsgroups/sci.crypt/15612
20_newsgroups/sci.crypt/15613
20_newsgroups/sci.crypt/15614
20_newsgroups/sci.crypt/15615
20_newsgroups/sci.crypt/15616
20_newsgroups/sci.crypt/15617
20_newsgroups/sci.crypt/15618
20_newsgroups/sci.crypt/15619
20_newsgroups/sci.crypt/15620
20_newsgro

20_newsgroups/sci.crypt/15976
20_newsgroups/sci.crypt/15977
20_newsgroups/sci.crypt/15978
20_newsgroups/sci.crypt/16077
20_newsgroups/sci.crypt/15980
20_newsgroups/sci.crypt/15981
20_newsgroups/sci.crypt/15982
20_newsgroups/sci.crypt/15983
20_newsgroups/sci.crypt/15984
20_newsgroups/sci.crypt/16025
20_newsgroups/sci.crypt/16041
20_newsgroups/sci.crypt/16042
20_newsgroups/sci.crypt/15985
20_newsgroups/sci.crypt/15986
20_newsgroups/sci.crypt/15987
20_newsgroups/sci.crypt/15988
20_newsgroups/sci.crypt/15989
20_newsgroups/sci.crypt/15990
20_newsgroups/sci.crypt/15991
20_newsgroups/sci.crypt/15992
20_newsgroups/sci.crypt/15993
20_newsgroups/sci.crypt/15994
20_newsgroups/sci.crypt/15995
20_newsgroups/sci.crypt/15996
20_newsgroups/sci.crypt/15997
20_newsgroups/sci.crypt/15998
20_newsgroups/sci.crypt/15999
20_newsgroups/sci.crypt/16000
20_newsgroups/sci.crypt/16001
20_newsgroups/sci.crypt/16002
20_newsgroups/sci.crypt/16003
20_newsgroups/sci.crypt/16004
20_newsgroups/sci.crypt/16005
20_newsgro

20_newsgroups/sci.electronics/53614
20_newsgroups/sci.electronics/53615
20_newsgroups/sci.electronics/53666
20_newsgroups/sci.electronics/53667
20_newsgroups/sci.electronics/53668
20_newsgroups/sci.electronics/53669
20_newsgroups/sci.electronics/53670
20_newsgroups/sci.electronics/53671
20_newsgroups/sci.electronics/53672
20_newsgroups/sci.electronics/53673
20_newsgroups/sci.electronics/53674
20_newsgroups/sci.electronics/53675
20_newsgroups/sci.electronics/53870
20_newsgroups/sci.electronics/53676
20_newsgroups/sci.electronics/53677
20_newsgroups/sci.electronics/53678
20_newsgroups/sci.electronics/53679
20_newsgroups/sci.electronics/53680
20_newsgroups/sci.electronics/53681
20_newsgroups/sci.electronics/53682
20_newsgroups/sci.electronics/53683
20_newsgroups/sci.electronics/53684
20_newsgroups/sci.electronics/53685
20_newsgroups/sci.electronics/53817
20_newsgroups/sci.electronics/53818
20_newsgroups/sci.electronics/53819
20_newsgroups/sci.electronics/53820
20_newsgroups/sci.electronic

20_newsgroups/sci.electronics/52722
20_newsgroups/sci.electronics/52723
20_newsgroups/sci.electronics/52724
20_newsgroups/sci.electronics/53699
20_newsgroups/sci.electronics/52725
20_newsgroups/sci.electronics/52726
20_newsgroups/sci.electronics/52727
20_newsgroups/sci.electronics/53700
20_newsgroups/sci.electronics/53701
20_newsgroups/sci.electronics/53702
20_newsgroups/sci.electronics/53703
20_newsgroups/sci.electronics/53704
20_newsgroups/sci.electronics/53705
20_newsgroups/sci.electronics/53706
20_newsgroups/sci.electronics/53707
20_newsgroups/sci.electronics/53708
20_newsgroups/sci.electronics/53709
20_newsgroups/sci.electronics/54023
20_newsgroups/sci.electronics/53710
20_newsgroups/sci.electronics/53711
20_newsgroups/sci.electronics/53712
20_newsgroups/sci.electronics/53713
20_newsgroups/sci.electronics/53714
20_newsgroups/sci.electronics/53715
20_newsgroups/sci.electronics/53716
20_newsgroups/sci.electronics/53717
20_newsgroups/sci.electronics/53718
20_newsgroups/sci.electronic

20_newsgroups/sci.electronics/53531
20_newsgroups/sci.electronics/53532
20_newsgroups/sci.electronics/53533
20_newsgroups/sci.electronics/53534
20_newsgroups/sci.electronics/53535
20_newsgroups/sci.electronics/53536
20_newsgroups/sci.electronics/53537
20_newsgroups/sci.electronics/53538
20_newsgroups/sci.electronics/52728
20_newsgroups/sci.electronics/52729
20_newsgroups/sci.electronics/52730
20_newsgroups/sci.electronics/52731
20_newsgroups/sci.electronics/52732
20_newsgroups/sci.electronics/52733
20_newsgroups/sci.electronics/52734
20_newsgroups/sci.electronics/52735
20_newsgroups/sci.electronics/52736
20_newsgroups/sci.electronics/52737
20_newsgroups/sci.electronics/52738
20_newsgroups/sci.electronics/52739
20_newsgroups/sci.electronics/52740
20_newsgroups/sci.electronics/52741
20_newsgroups/sci.electronics/52742
20_newsgroups/sci.electronics/52743
20_newsgroups/sci.electronics/52744
20_newsgroups/sci.electronics/52745
20_newsgroups/sci.electronics/52746
20_newsgroups/sci.electronic

20_newsgroups/sci.electronics/54319
20_newsgroups/sci.electronics/54320
20_newsgroups/sci.electronics/54321
20_newsgroups/sci.electronics/54322
20_newsgroups/sci.electronics/54323
20_newsgroups/sci.electronics/54324
20_newsgroups/sci.electronics/54325
20_newsgroups/sci.electronics/54326
20_newsgroups/sci.electronics/54327
20_newsgroups/sci.electronics/54328
20_newsgroups/sci.electronics/54329
20_newsgroups/sci.electronics/54330
20_newsgroups/sci.electronics/54331
20_newsgroups/sci.electronics/54332
20_newsgroups/sci.electronics/54333
20_newsgroups/sci.electronics/54339
20_newsgroups/sci.electronics/54340
20_newsgroups/sci.electronics/54341
20_newsgroups/sci.electronics/54342
20_newsgroups/sci.electronics/54343
20_newsgroups/sci.electronics/54346
20_newsgroups/sci.electronics/54347
20_newsgroups/sci.electronics/54348
20_newsgroups/sci.electronics/54349
20_newsgroups/sci.electronics/54350
20_newsgroups/sci.electronics/54351
20_newsgroups/sci.electronics/54240
20_newsgroups/sci.electronic

20_newsgroups/sci.med/59124
20_newsgroups/sci.med/59125
20_newsgroups/sci.med/59126
20_newsgroups/sci.med/59127
20_newsgroups/sci.med/59128
20_newsgroups/sci.med/58902
20_newsgroups/sci.med/59129
20_newsgroups/sci.med/59130
20_newsgroups/sci.med/59131
20_newsgroups/sci.med/59132
20_newsgroups/sci.med/58903
20_newsgroups/sci.med/59133
20_newsgroups/sci.med/59134
20_newsgroups/sci.med/59135
20_newsgroups/sci.med/59136
20_newsgroups/sci.med/59137
20_newsgroups/sci.med/59138
20_newsgroups/sci.med/59139
20_newsgroups/sci.med/59140
20_newsgroups/sci.med/59141
20_newsgroups/sci.med/59142
20_newsgroups/sci.med/59143
20_newsgroups/sci.med/59144
20_newsgroups/sci.med/59145
20_newsgroups/sci.med/59146
20_newsgroups/sci.med/59147
20_newsgroups/sci.med/59148
20_newsgroups/sci.med/59149
20_newsgroups/sci.med/59150
20_newsgroups/sci.med/59151
20_newsgroups/sci.med/59152
20_newsgroups/sci.med/59153
20_newsgroups/sci.med/59154
20_newsgroups/sci.med/59155
20_newsgroups/sci.med/59156
20_newsgroups/sci.me

20_newsgroups/sci.med/58860
20_newsgroups/sci.med/58861
20_newsgroups/sci.med/58862
20_newsgroups/sci.med/58863
20_newsgroups/sci.med/58864
20_newsgroups/sci.med/58865
20_newsgroups/sci.med/58866
20_newsgroups/sci.med/58867
20_newsgroups/sci.med/58868
20_newsgroups/sci.med/58869
20_newsgroups/sci.med/58870
20_newsgroups/sci.med/58871
20_newsgroups/sci.med/58872
20_newsgroups/sci.med/58873
20_newsgroups/sci.med/58874
20_newsgroups/sci.med/58875
20_newsgroups/sci.med/58876
20_newsgroups/sci.med/58877
20_newsgroups/sci.med/58878
20_newsgroups/sci.med/58879
20_newsgroups/sci.med/58880
20_newsgroups/sci.med/58881
20_newsgroups/sci.med/58882
20_newsgroups/sci.med/58883
20_newsgroups/sci.med/58120
20_newsgroups/sci.med/58121
20_newsgroups/sci.med/58122
20_newsgroups/sci.med/58123
20_newsgroups/sci.med/58124
20_newsgroups/sci.med/58125
20_newsgroups/sci.med/58126
20_newsgroups/sci.med/58127
20_newsgroups/sci.med/58128
20_newsgroups/sci.med/58129
20_newsgroups/sci.med/58130
20_newsgroups/sci.me

20_newsgroups/sci.med/59430
20_newsgroups/sci.med/59431
20_newsgroups/sci.med/59432
20_newsgroups/sci.med/59433
20_newsgroups/sci.med/59434
20_newsgroups/sci.med/59435
20_newsgroups/sci.med/59436
20_newsgroups/sci.med/59437
20_newsgroups/sci.med/59438
20_newsgroups/sci.med/59439
20_newsgroups/sci.med/59440
20_newsgroups/sci.med/59441
20_newsgroups/sci.med/59442
20_newsgroups/sci.med/59443
20_newsgroups/sci.med/59444
20_newsgroups/sci.med/59445
20_newsgroups/sci.med/59446
20_newsgroups/sci.med/59485
20_newsgroups/sci.med/59486
20_newsgroups/sci.med/59487
20_newsgroups/sci.med/59488
20_newsgroups/sci.med/59489
20_newsgroups/sci.med/59490
20_newsgroups/sci.med/59491
20_newsgroups/sci.med/59492
20_newsgroups/sci.med/59493
20_newsgroups/sci.med/59494
20_newsgroups/sci.med/59495
20_newsgroups/sci.med/59496
20_newsgroups/sci.med/59497
20_newsgroups/sci.med/59498
20_newsgroups/sci.med/59499
20_newsgroups/sci.med/59500
20_newsgroups/sci.med/59501
20_newsgroups/sci.med/59502
20_newsgroups/sci.me

20_newsgroups/sci.space/60154
20_newsgroups/sci.space/60155
20_newsgroups/sci.space/60156
20_newsgroups/sci.space/60157
20_newsgroups/sci.space/60158
20_newsgroups/sci.space/60159
20_newsgroups/sci.space/60160
20_newsgroups/sci.space/60161
20_newsgroups/sci.space/60867
20_newsgroups/sci.space/60868
20_newsgroups/sci.space/60869
20_newsgroups/sci.space/60870
20_newsgroups/sci.space/60871
20_newsgroups/sci.space/60872
20_newsgroups/sci.space/60873
20_newsgroups/sci.space/60874
20_newsgroups/sci.space/60875
20_newsgroups/sci.space/60876
20_newsgroups/sci.space/60877
20_newsgroups/sci.space/60878
20_newsgroups/sci.space/60879
20_newsgroups/sci.space/60880
20_newsgroups/sci.space/60881
20_newsgroups/sci.space/60882
20_newsgroups/sci.space/60883
20_newsgroups/sci.space/60884
20_newsgroups/sci.space/60885
20_newsgroups/sci.space/60886
20_newsgroups/sci.space/60887
20_newsgroups/sci.space/60888
20_newsgroups/sci.space/60889
20_newsgroups/sci.space/60890
20_newsgroups/sci.space/60891
20_newsgro

20_newsgroups/sci.space/61232
20_newsgroups/sci.space/61233
20_newsgroups/sci.space/61234
20_newsgroups/sci.space/61173
20_newsgroups/sci.space/61174
20_newsgroups/sci.space/61175
20_newsgroups/sci.space/61176
20_newsgroups/sci.space/61235
20_newsgroups/sci.space/61245
20_newsgroups/sci.space/61247
20_newsgroups/sci.space/61248
20_newsgroups/sci.space/61249
20_newsgroups/sci.space/61250
20_newsgroups/sci.space/61251
20_newsgroups/sci.space/61252
20_newsgroups/sci.space/61253
20_newsgroups/sci.space/61254
20_newsgroups/sci.space/61177
20_newsgroups/sci.space/61178
20_newsgroups/sci.space/61179
20_newsgroups/sci.space/61236
20_newsgroups/sci.space/61180
20_newsgroups/sci.space/61237
20_newsgroups/sci.space/61238
20_newsgroups/sci.space/61239
20_newsgroups/sci.space/61240
20_newsgroups/sci.space/61255
20_newsgroups/sci.space/61256
20_newsgroups/sci.space/61257
20_newsgroups/sci.space/61258
20_newsgroups/sci.space/61259
20_newsgroups/sci.space/61260
20_newsgroups/sci.space/61261
20_newsgro

20_newsgroups/sci.space/62450
20_newsgroups/sci.space/62451
20_newsgroups/sci.space/62419
20_newsgroups/sci.space/62452
20_newsgroups/sci.space/62453
20_newsgroups/sci.space/62454
20_newsgroups/sci.space/62455
20_newsgroups/sci.space/62474
20_newsgroups/sci.space/62104
20_newsgroups/sci.space/62614
20_newsgroups/sci.space/62615
20_newsgroups/sci.space/62475
20_newsgroups/sci.space/62709
20_newsgroups/sci.space/62476
20_newsgroups/sci.space/62317
20_newsgroups/sci.space/62318
20_newsgroups/sci.space/62319
20_newsgroups/sci.space/62420
20_newsgroups/sci.space/62477
20_newsgroups/sci.space/62478
20_newsgroups/sci.space/62479
20_newsgroups/sci.space/62480
20_newsgroups/sci.space/62481
20_newsgroups/sci.space/62616
20_newsgroups/sci.space/62116
20_newsgroups/sci.space/62117
20_newsgroups/sci.space/62118
20_newsgroups/sci.space/62126
20_newsgroups/sci.space/62127
20_newsgroups/sci.space/62129
20_newsgroups/sci.space/62130
20_newsgroups/sci.space/62131
20_newsgroups/soc.religion.christian/
20

20_newsgroups/soc.religion.christian/20875
20_newsgroups/soc.religion.christian/20876
20_newsgroups/soc.religion.christian/20877
20_newsgroups/soc.religion.christian/20878
20_newsgroups/soc.religion.christian/20879
20_newsgroups/soc.religion.christian/20880
20_newsgroups/soc.religion.christian/20881
20_newsgroups/soc.religion.christian/20704
20_newsgroups/soc.religion.christian/20706
20_newsgroups/soc.religion.christian/20707
20_newsgroups/soc.religion.christian/20708
20_newsgroups/soc.religion.christian/20709
20_newsgroups/soc.religion.christian/20710
20_newsgroups/soc.religion.christian/20711
20_newsgroups/soc.religion.christian/20712
20_newsgroups/soc.religion.christian/20713
20_newsgroups/soc.religion.christian/20714
20_newsgroups/soc.religion.christian/20715
20_newsgroups/soc.religion.christian/20716
20_newsgroups/soc.religion.christian/20717
20_newsgroups/soc.religion.christian/20718
20_newsgroups/soc.religion.christian/20719
20_newsgroups/soc.religion.christian/20720
20_newsgrou

20_newsgroups/soc.religion.christian/20912
20_newsgroups/soc.religion.christian/20913
20_newsgroups/soc.religion.christian/20914
20_newsgroups/soc.religion.christian/20915
20_newsgroups/soc.religion.christian/20916
20_newsgroups/soc.religion.christian/20917
20_newsgroups/soc.religion.christian/20918
20_newsgroups/soc.religion.christian/20919
20_newsgroups/soc.religion.christian/20920
20_newsgroups/soc.religion.christian/20921
20_newsgroups/soc.religion.christian/20922
20_newsgroups/soc.religion.christian/20923
20_newsgroups/soc.religion.christian/20924
20_newsgroups/soc.religion.christian/20925
20_newsgroups/soc.religion.christian/20926
20_newsgroups/soc.religion.christian/20927
20_newsgroups/soc.religion.christian/20928
20_newsgroups/soc.religion.christian/20929
20_newsgroups/soc.religion.christian/20930
20_newsgroups/soc.religion.christian/20931
20_newsgroups/soc.religion.christian/20932
20_newsgroups/soc.religion.christian/20933
20_newsgroups/soc.religion.christian/20934
20_newsgrou

20_newsgroups/soc.religion.christian/21478
20_newsgroups/soc.religion.christian/21479
20_newsgroups/soc.religion.christian/21480
20_newsgroups/soc.religion.christian/21481
20_newsgroups/soc.religion.christian/21482
20_newsgroups/soc.religion.christian/21483
20_newsgroups/soc.religion.christian/21484
20_newsgroups/soc.religion.christian/21485
20_newsgroups/soc.religion.christian/21486
20_newsgroups/soc.religion.christian/21487
20_newsgroups/soc.religion.christian/21488
20_newsgroups/soc.religion.christian/21489
20_newsgroups/soc.religion.christian/21490
20_newsgroups/soc.religion.christian/21491
20_newsgroups/soc.religion.christian/21492
20_newsgroups/soc.religion.christian/21493
20_newsgroups/soc.religion.christian/21494
20_newsgroups/soc.religion.christian/21495
20_newsgroups/soc.religion.christian/21496
20_newsgroups/soc.religion.christian/21532
20_newsgroups/soc.religion.christian/21533
20_newsgroups/soc.religion.christian/21534
20_newsgroups/soc.religion.christian/21719
20_newsgrou

20_newsgroups/soc.religion.christian/21732
20_newsgroups/soc.religion.christian/21733
20_newsgroups/soc.religion.christian/21734
20_newsgroups/soc.religion.christian/21735
20_newsgroups/soc.religion.christian/21736
20_newsgroups/soc.religion.christian/21737
20_newsgroups/soc.religion.christian/21738
20_newsgroups/soc.religion.christian/21739
20_newsgroups/soc.religion.christian/21740
20_newsgroups/soc.religion.christian/21741
20_newsgroups/soc.religion.christian/21742
20_newsgroups/soc.religion.christian/21743
20_newsgroups/soc.religion.christian/21744
20_newsgroups/soc.religion.christian/21745
20_newsgroups/soc.religion.christian/21746
20_newsgroups/soc.religion.christian/21747
20_newsgroups/soc.religion.christian/21748
20_newsgroups/soc.religion.christian/21749
20_newsgroups/soc.religion.christian/21750
20_newsgroups/soc.religion.christian/21751
20_newsgroups/soc.religion.christian/21752
20_newsgroups/soc.religion.christian/21753
20_newsgroups/soc.religion.christian/21754
20_newsgrou

20_newsgroups/talk.politics.guns/54205
20_newsgroups/talk.politics.guns/54206
20_newsgroups/talk.politics.guns/54207
20_newsgroups/talk.politics.guns/54208
20_newsgroups/talk.politics.guns/54209
20_newsgroups/talk.politics.guns/54210
20_newsgroups/talk.politics.guns/54211
20_newsgroups/talk.politics.guns/54212
20_newsgroups/talk.politics.guns/54213
20_newsgroups/talk.politics.guns/54214
20_newsgroups/talk.politics.guns/54215
20_newsgroups/talk.politics.guns/54216
20_newsgroups/talk.politics.guns/54217
20_newsgroups/talk.politics.guns/54218
20_newsgroups/talk.politics.guns/54219
20_newsgroups/talk.politics.guns/54220
20_newsgroups/talk.politics.guns/54221
20_newsgroups/talk.politics.guns/54222
20_newsgroups/talk.politics.guns/54223
20_newsgroups/talk.politics.guns/54224
20_newsgroups/talk.politics.guns/54225
20_newsgroups/talk.politics.guns/54226
20_newsgroups/talk.politics.guns/54233
20_newsgroups/talk.politics.guns/54234
20_newsgroups/talk.politics.guns/54235
20_newsgroups/talk.politi

20_newsgroups/talk.politics.guns/54245
20_newsgroups/talk.politics.guns/54246
20_newsgroups/talk.politics.guns/54247
20_newsgroups/talk.politics.guns/54248
20_newsgroups/talk.politics.guns/54249
20_newsgroups/talk.politics.guns/54250
20_newsgroups/talk.politics.guns/54251
20_newsgroups/talk.politics.guns/54252
20_newsgroups/talk.politics.guns/54253
20_newsgroups/talk.politics.guns/54254
20_newsgroups/talk.politics.guns/54255
20_newsgroups/talk.politics.guns/54256
20_newsgroups/talk.politics.guns/54257
20_newsgroups/talk.politics.guns/54258
20_newsgroups/talk.politics.guns/54259
20_newsgroups/talk.politics.guns/54260
20_newsgroups/talk.politics.guns/54261
20_newsgroups/talk.politics.guns/54262
20_newsgroups/talk.politics.guns/54263
20_newsgroups/talk.politics.guns/54264
20_newsgroups/talk.politics.guns/54265
20_newsgroups/talk.politics.guns/54266
20_newsgroups/talk.politics.guns/54267
20_newsgroups/talk.politics.guns/54268
20_newsgroups/talk.politics.guns/54269
20_newsgroups/talk.politi

20_newsgroups/talk.politics.guns/54603
20_newsgroups/talk.politics.guns/54604
20_newsgroups/talk.politics.guns/54605
20_newsgroups/talk.politics.guns/54608
20_newsgroups/talk.politics.guns/54609
20_newsgroups/talk.politics.guns/54610
20_newsgroups/talk.politics.guns/54611
20_newsgroups/talk.politics.guns/54612
20_newsgroups/talk.politics.guns/54613
20_newsgroups/talk.politics.guns/54614
20_newsgroups/talk.politics.guns/54615
20_newsgroups/talk.politics.guns/54616
20_newsgroups/talk.politics.guns/54617
20_newsgroups/talk.politics.guns/54618
20_newsgroups/talk.politics.guns/54619
20_newsgroups/talk.politics.guns/54620
20_newsgroups/talk.politics.guns/54621
20_newsgroups/talk.politics.guns/54622
20_newsgroups/talk.politics.guns/54623
20_newsgroups/talk.politics.guns/54624
20_newsgroups/talk.politics.guns/54625
20_newsgroups/talk.politics.guns/54626
20_newsgroups/talk.politics.guns/54627
20_newsgroups/talk.politics.guns/54628
20_newsgroups/talk.politics.guns/54629
20_newsgroups/talk.politi

20_newsgroups/talk.politics.guns/54880
20_newsgroups/talk.politics.guns/54881
20_newsgroups/talk.politics.guns/55058
20_newsgroups/talk.politics.guns/55059
20_newsgroups/talk.politics.guns/54882
20_newsgroups/talk.politics.guns/54883
20_newsgroups/talk.politics.guns/54884
20_newsgroups/talk.politics.guns/54885
20_newsgroups/talk.politics.guns/54886
20_newsgroups/talk.politics.guns/54887
20_newsgroups/talk.politics.guns/54888
20_newsgroups/talk.politics.guns/54889
20_newsgroups/talk.politics.guns/54890
20_newsgroups/talk.politics.guns/54891
20_newsgroups/talk.politics.guns/54892
20_newsgroups/talk.politics.guns/54893
20_newsgroups/talk.politics.guns/54894
20_newsgroups/talk.politics.guns/54895
20_newsgroups/talk.politics.guns/54896
20_newsgroups/talk.politics.guns/54897
20_newsgroups/talk.politics.guns/54898
20_newsgroups/talk.politics.guns/54899
20_newsgroups/talk.politics.guns/54900
20_newsgroups/talk.politics.guns/54903
20_newsgroups/talk.politics.guns/54904
20_newsgroups/talk.politi

20_newsgroups/talk.politics.mideast/75369
20_newsgroups/talk.politics.mideast/76122
20_newsgroups/talk.politics.mideast/76123
20_newsgroups/talk.politics.mideast/76124
20_newsgroups/talk.politics.mideast/76131
20_newsgroups/talk.politics.mideast/75885
20_newsgroups/talk.politics.mideast/75886
20_newsgroups/talk.politics.mideast/76132
20_newsgroups/talk.politics.mideast/76133
20_newsgroups/talk.politics.mideast/75887
20_newsgroups/talk.politics.mideast/75888
20_newsgroups/talk.politics.mideast/75889
20_newsgroups/talk.politics.mideast/75890
20_newsgroups/talk.politics.mideast/75891
20_newsgroups/talk.politics.mideast/75951
20_newsgroups/talk.politics.mideast/76134
20_newsgroups/talk.politics.mideast/75892
20_newsgroups/talk.politics.mideast/75893
20_newsgroups/talk.politics.mideast/75894
20_newsgroups/talk.politics.mideast/75895
20_newsgroups/talk.politics.mideast/76135
20_newsgroups/talk.politics.mideast/76136
20_newsgroups/talk.politics.mideast/76137
20_newsgroups/talk.politics.mideas

20_newsgroups/talk.politics.mideast/75975
20_newsgroups/talk.politics.mideast/75976
20_newsgroups/talk.politics.mideast/75977
20_newsgroups/talk.politics.mideast/75978
20_newsgroups/talk.politics.mideast/75979
20_newsgroups/talk.politics.mideast/75980
20_newsgroups/talk.politics.mideast/75981
20_newsgroups/talk.politics.mideast/75982
20_newsgroups/talk.politics.mideast/75983
20_newsgroups/talk.politics.mideast/75984
20_newsgroups/talk.politics.mideast/75985
20_newsgroups/talk.politics.mideast/75986
20_newsgroups/talk.politics.mideast/75987
20_newsgroups/talk.politics.mideast/75988
20_newsgroups/talk.politics.mideast/75989
20_newsgroups/talk.politics.mideast/75990
20_newsgroups/talk.politics.mideast/75991
20_newsgroups/talk.politics.mideast/75993
20_newsgroups/talk.politics.mideast/75994
20_newsgroups/talk.politics.mideast/75995
20_newsgroups/talk.politics.mideast/75996
20_newsgroups/talk.politics.mideast/75997
20_newsgroups/talk.politics.mideast/75998
20_newsgroups/talk.politics.mideas

20_newsgroups/talk.politics.mideast/76257
20_newsgroups/talk.politics.mideast/76258
20_newsgroups/talk.politics.mideast/76259
20_newsgroups/talk.politics.mideast/76260
20_newsgroups/talk.politics.mideast/76261
20_newsgroups/talk.politics.mideast/76262
20_newsgroups/talk.politics.mideast/76263
20_newsgroups/talk.politics.mideast/76264
20_newsgroups/talk.politics.mideast/76265
20_newsgroups/talk.politics.mideast/76266
20_newsgroups/talk.politics.mideast/76267
20_newsgroups/talk.politics.mideast/76268
20_newsgroups/talk.politics.mideast/76269
20_newsgroups/talk.politics.mideast/76270
20_newsgroups/talk.politics.mideast/76293
20_newsgroups/talk.politics.mideast/76294
20_newsgroups/talk.politics.mideast/76295
20_newsgroups/talk.politics.mideast/76296
20_newsgroups/talk.politics.mideast/76297
20_newsgroups/talk.politics.mideast/76298
20_newsgroups/talk.politics.mideast/76299
20_newsgroups/talk.politics.mideast/76300
20_newsgroups/talk.politics.mideast/76301
20_newsgroups/talk.politics.mideas

20_newsgroups/talk.politics.mideast/77328
20_newsgroups/talk.politics.mideast/77329
20_newsgroups/talk.politics.mideast/77330
20_newsgroups/talk.politics.mideast/77337
20_newsgroups/talk.politics.mideast/77338
20_newsgroups/talk.politics.mideast/77339
20_newsgroups/talk.politics.mideast/77340
20_newsgroups/talk.politics.mideast/77341
20_newsgroups/talk.politics.mideast/77342
20_newsgroups/talk.politics.mideast/77190
20_newsgroups/talk.politics.mideast/77343
20_newsgroups/talk.politics.mideast/77191
20_newsgroups/talk.politics.mideast/77192
20_newsgroups/talk.politics.mideast/77193
20_newsgroups/talk.politics.mideast/77194
20_newsgroups/talk.politics.mideast/77344
20_newsgroups/talk.politics.mideast/77345
20_newsgroups/talk.politics.mideast/77210
20_newsgroups/talk.politics.mideast/77211
20_newsgroups/talk.politics.mideast/77355
20_newsgroups/talk.politics.mideast/77356
20_newsgroups/talk.politics.mideast/77357
20_newsgroups/talk.politics.mideast/77358
20_newsgroups/talk.politics.mideas

20_newsgroups/talk.politics.mideast/77814
20_newsgroups/talk.politics.mideast/77815
20_newsgroups/talk.politics.mideast/77816
20_newsgroups/talk.politics.mideast/77850
20_newsgroups/talk.politics.mideast/77817
20_newsgroups/talk.politics.mideast/77818
20_newsgroups/talk.politics.mideast/77819
20_newsgroups/talk.politics.mideast/77820
20_newsgroups/talk.politics.mideast/77821
20_newsgroups/talk.politics.mideast/77822
20_newsgroups/talk.politics.mideast/77823
20_newsgroups/talk.politics.misc/
20_newsgroups/talk.politics.misc/178293
20_newsgroups/talk.politics.misc/179034
20_newsgroups/talk.politics.misc/179040
20_newsgroups/talk.politics.misc/179041
20_newsgroups/talk.politics.misc/179042
20_newsgroups/talk.politics.misc/179046
20_newsgroups/talk.politics.misc/179047
20_newsgroups/talk.politics.misc/179048
20_newsgroups/talk.politics.misc/179049
20_newsgroups/talk.politics.misc/178294
20_newsgroups/talk.politics.misc/179050
20_newsgroups/talk.politics.misc/179053
20_newsgroups/talk.polit

20_newsgroups/talk.politics.misc/178431
20_newsgroups/talk.politics.misc/178432
20_newsgroups/talk.politics.misc/178433
20_newsgroups/talk.politics.misc/178434
20_newsgroups/talk.politics.misc/178435
20_newsgroups/talk.politics.misc/178436
20_newsgroups/talk.politics.misc/178437
20_newsgroups/talk.politics.misc/178438
20_newsgroups/talk.politics.misc/178439
20_newsgroups/talk.politics.misc/178440
20_newsgroups/talk.politics.misc/178441
20_newsgroups/talk.politics.misc/178442
20_newsgroups/talk.politics.misc/178443
20_newsgroups/talk.politics.misc/178444
20_newsgroups/talk.politics.misc/178445
20_newsgroups/talk.politics.misc/178446
20_newsgroups/talk.politics.misc/178447
20_newsgroups/talk.politics.misc/178448
20_newsgroups/talk.politics.misc/178449
20_newsgroups/talk.politics.misc/178450
20_newsgroups/talk.politics.misc/178451
20_newsgroups/talk.politics.misc/178452
20_newsgroups/talk.politics.misc/178453
20_newsgroups/talk.politics.misc/178454
20_newsgroups/talk.politics.misc/178455


20_newsgroups/talk.politics.misc/178818
20_newsgroups/talk.politics.misc/178819
20_newsgroups/talk.politics.misc/178820
20_newsgroups/talk.politics.misc/178821
20_newsgroups/talk.politics.misc/178822
20_newsgroups/talk.politics.misc/178823
20_newsgroups/talk.politics.misc/178824
20_newsgroups/talk.politics.misc/178825
20_newsgroups/talk.politics.misc/178826
20_newsgroups/talk.politics.misc/178827
20_newsgroups/talk.politics.misc/178828
20_newsgroups/talk.politics.misc/178829
20_newsgroups/talk.politics.misc/178830
20_newsgroups/talk.politics.misc/178831
20_newsgroups/talk.politics.misc/178832
20_newsgroups/talk.politics.misc/178833
20_newsgroups/talk.politics.misc/178834
20_newsgroups/talk.politics.misc/178835
20_newsgroups/talk.politics.misc/178836
20_newsgroups/talk.politics.misc/178837
20_newsgroups/talk.politics.misc/178838
20_newsgroups/talk.politics.misc/178839
20_newsgroups/talk.politics.misc/178840
20_newsgroups/talk.politics.misc/178841
20_newsgroups/talk.politics.misc/178842


20_newsgroups/talk.politics.misc/178575
20_newsgroups/talk.politics.misc/178576
20_newsgroups/talk.politics.misc/178577
20_newsgroups/talk.politics.misc/178578
20_newsgroups/talk.politics.misc/178579
20_newsgroups/talk.politics.misc/178591
20_newsgroups/talk.politics.misc/178581
20_newsgroups/talk.politics.misc/178582
20_newsgroups/talk.politics.misc/178583
20_newsgroups/talk.politics.misc/178584
20_newsgroups/talk.politics.misc/178585
20_newsgroups/talk.politics.misc/178586
20_newsgroups/talk.politics.misc/178587
20_newsgroups/talk.politics.misc/178588
20_newsgroups/talk.politics.misc/178589
20_newsgroups/talk.politics.misc/178590
20_newsgroups/talk.politics.misc/178592
20_newsgroups/talk.politics.misc/178593
20_newsgroups/talk.politics.misc/178594
20_newsgroups/talk.politics.misc/178595
20_newsgroups/talk.politics.misc/178596
20_newsgroups/talk.politics.misc/178597
20_newsgroups/talk.politics.misc/178598
20_newsgroups/talk.politics.misc/178601
20_newsgroups/talk.politics.misc/178602


20_newsgroups/talk.politics.misc/179084
20_newsgroups/talk.politics.misc/179085
20_newsgroups/talk.politics.misc/179086
20_newsgroups/talk.politics.misc/179087
20_newsgroups/talk.politics.misc/179088
20_newsgroups/talk.politics.misc/179089
20_newsgroups/talk.politics.misc/179090
20_newsgroups/talk.politics.misc/179091
20_newsgroups/talk.politics.misc/179092
20_newsgroups/talk.politics.misc/179097
20_newsgroups/talk.politics.misc/179098
20_newsgroups/talk.politics.misc/179099
20_newsgroups/talk.politics.misc/179100
20_newsgroups/talk.politics.misc/179101
20_newsgroups/talk.politics.misc/179102
20_newsgroups/talk.politics.misc/179103
20_newsgroups/talk.politics.misc/179104
20_newsgroups/talk.politics.misc/179105
20_newsgroups/talk.politics.misc/179106
20_newsgroups/talk.politics.misc/179107
20_newsgroups/talk.politics.misc/179108
20_newsgroups/talk.politics.misc/179109
20_newsgroups/talk.politics.misc/179110
20_newsgroups/talk.politics.misc/179111
20_newsgroups/talk.politics.misc/179112


20_newsgroups/talk.religion.misc/83772
20_newsgroups/talk.religion.misc/83773
20_newsgroups/talk.religion.misc/83774
20_newsgroups/talk.religion.misc/83775
20_newsgroups/talk.religion.misc/83776
20_newsgroups/talk.religion.misc/83777
20_newsgroups/talk.religion.misc/83778
20_newsgroups/talk.religion.misc/83779
20_newsgroups/talk.religion.misc/83780
20_newsgroups/talk.religion.misc/83781
20_newsgroups/talk.religion.misc/83782
20_newsgroups/talk.religion.misc/83783
20_newsgroups/talk.religion.misc/83784
20_newsgroups/talk.religion.misc/83785
20_newsgroups/talk.religion.misc/83786
20_newsgroups/talk.religion.misc/83787
20_newsgroups/talk.religion.misc/83788
20_newsgroups/talk.religion.misc/83789
20_newsgroups/talk.religion.misc/83790
20_newsgroups/talk.religion.misc/83791
20_newsgroups/talk.religion.misc/83881
20_newsgroups/talk.religion.misc/83882
20_newsgroups/talk.religion.misc/83883
20_newsgroups/talk.religion.misc/83884
20_newsgroups/talk.religion.misc/83885
20_newsgroups/talk.religi

20_newsgroups/talk.religion.misc/83688
20_newsgroups/talk.religion.misc/83689
20_newsgroups/talk.religion.misc/83690
20_newsgroups/talk.religion.misc/83691
20_newsgroups/talk.religion.misc/83692
20_newsgroups/talk.religion.misc/83693
20_newsgroups/talk.religion.misc/83694
20_newsgroups/talk.religion.misc/83721
20_newsgroups/talk.religion.misc/83722
20_newsgroups/talk.religion.misc/83723
20_newsgroups/talk.religion.misc/83724
20_newsgroups/talk.religion.misc/83725
20_newsgroups/talk.religion.misc/83726
20_newsgroups/talk.religion.misc/83727
20_newsgroups/talk.religion.misc/83728
20_newsgroups/talk.religion.misc/83729
20_newsgroups/talk.religion.misc/83730
20_newsgroups/talk.religion.misc/83731
20_newsgroups/talk.religion.misc/83732
20_newsgroups/talk.religion.misc/83733
20_newsgroups/talk.religion.misc/83743
20_newsgroups/talk.religion.misc/83744
20_newsgroups/talk.religion.misc/83745
20_newsgroups/talk.religion.misc/83746
20_newsgroups/talk.religion.misc/83747
20_newsgroups/talk.religi

20_newsgroups/talk.religion.misc/84067
20_newsgroups/talk.religion.misc/84068
20_newsgroups/talk.religion.misc/84069
20_newsgroups/talk.religion.misc/84070
20_newsgroups/talk.religion.misc/84071
20_newsgroups/talk.religion.misc/84072
20_newsgroups/talk.religion.misc/84073
20_newsgroups/talk.religion.misc/84074
20_newsgroups/talk.religion.misc/84075
20_newsgroups/talk.religion.misc/84076
20_newsgroups/talk.religion.misc/84077
20_newsgroups/talk.religion.misc/84078
20_newsgroups/talk.religion.misc/84079
20_newsgroups/talk.religion.misc/84080
20_newsgroups/talk.religion.misc/84081
20_newsgroups/talk.religion.misc/84082
20_newsgroups/talk.religion.misc/84083
20_newsgroups/talk.religion.misc/84084
20_newsgroups/talk.religion.misc/84085
20_newsgroups/talk.religion.misc/84086
20_newsgroups/talk.religion.misc/84087
20_newsgroups/talk.religion.misc/84175
20_newsgroups/talk.religion.misc/84088
20_newsgroups/talk.religion.misc/84090
20_newsgroups/talk.religion.misc/84091
20_newsgroups/talk.religi

20_newsgroups/talk.religion.misc/84357
20_newsgroups/talk.religion.misc/84358
20_newsgroups/talk.religion.misc/84359
20_newsgroups/talk.religion.misc/84360
rm: cannot remove '20news-19997*': No such file or directory
directories in 20 newsgroup dataset


listing = 


"alt.atheism
"


alt.atheism
comp.graphics
comp.os.ms-windows.misc
comp.sys.ibm.pc.hardware
comp.sys.mac.hardware
comp.windows.x
misc.forsale
rec.autos
rec.motorcycles
rec.sport.baseball
rec.sport.hockey
sci.crypt
sci.electronics
sci.med
sci.space
soc.religion.christian
talk.politics.guns
talk.politics.mideast
talk.politics.misc
talk.religion.misc


We will study the full dataset, but if you want to run things a bit faster you can sample only a fraction of the dataset.  The embeddings and clusters will be different, however.

In [3]:
val corpusFraction = 1
val seed = 0L
val corpusFull: RDD[String] = sc.wholeTextFiles("20_newsgroups/*/*").map(_._2)

val corpus = if (corpusFraction == 1.0){corpusFull} else{
               corpusFull.sample(fraction = corpusFraction, withReplacement = false, seed = seed)}

corpusFraction = 1
seed = 0
corpusFull = MapPartitionsRDD[2] at map at <console>:49
corpus = MapPartitionsRDD[2] at map at <console>:49


MapPartitionsRDD[2] at map at <console>:49

The first document of the corpus if from the `talk.politics.misc directory`, with the filename 179112 

In [4]:
corpus.take(1)

Array("Xref: cantaloupe.srv.cs.cmu.edu misc.legal:61051 talk.politics.misc:179112
Path: cantaloupe.srv.cs.cmu.edu!magnesium.club.cc.cmu.edu!news.sei.cmu.edu!cis.ohio-state.edu!zaphod.mps.ohio-state.edu!cs.utexas.edu!uunet!optilink!cramer
From: cramer@optilink.COM (Clayton Cramer)
Newsgroups: misc.legal,talk.politics.misc
Subject: Re: WACO: Clinton press conference, part 1
Message-ID: <15577@optilink.COM>
Date: 23 Apr 93 21:47:27 GMT
References: <1993Apr21.021301.25113@r-node.hub.org> <93Apr22.234553edt.47633@neat.cs.toronto.edu>
Organization: Optilink Corporation, Petaluma, CA
Lines: 10
In article <93Apr22.234553edt.47633@neat.cs.toronto.edu>, quoctp@cs.toronto.edu (Quoc Tuan Pham) writes:
> Did anyone notice that Clinton was smiling and making jokes just before
...


[Xref: cantaloupe.srv.cs.cmu.edu misc.legal:61051 talk.politics.misc:179112
Path: cantaloupe.srv.cs.cmu.edu!magnesium.club.cc.cmu.edu!news.sei.cmu.edu!cis.ohio-state.edu!zaphod.mps.ohio-state.edu!cs.utexas.edu!uunet!optilink!cramer
From: cramer@optilink.COM (Clayton Cramer)
Newsgroups: misc.legal,talk.politics.misc
Subject: Re: WACO: Clinton press conference, part 1
Message-ID: <15577@optilink.COM>
Date: 23 Apr 93 21:47:27 GMT
References: <1993Apr21.021301.25113@r-node.hub.org> <93Apr22.234553edt.47633@neat.cs.toronto.edu>
Organization: Optilink Corporation, Petaluma, CA
Lines: 10

In article <93Apr22.234553edt.47633@neat.cs.toronto.edu>, quoctp@cs.toronto.edu (Quoc Tuan Pham) writes:
> Did anyone notice that Clinton was smiling and making jokes just before
> this press conference? Considering the number of people killed, this 
> seems very inappropriate to me.

Why?  He, Reno, and the FBI got what they wanted -- a reminder of
who is the boss in America -- the thugs who work for the go

In [5]:
//corpus.persist()

Name: Syntax Error.
Message: 
StackTrace: 

# Data Preparation

First we filter out the small words (<3 letters), and then we convert the corpus into a DataFrame, which is what the ML library uses.

In [6]:
val documentDF = corpus.map(_.toLowerCase.split("\\s")).
                        map(_.filter(_.length > 3).filter(_.forall(java.lang.Character.isLetter))).
                        toDF("text")

documentDF = [text: array<string>]


[text: array<string>]

Here is what the first article looks like after preparation.

In [7]:
documentDF.select("text").show(1,false)

+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|text                                                                                                                                                                                                                                                                                                           |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|[clinton, press, part, optilink, article, tuan, anyone, notice, that, clinton, sm

## Word2Vec Training

The training of the Word2Vec model is more compute intensive than the actual clustering.  This is because the embedding uses a neural network.  For now, we will consider this as a mapping between groups of words and a vector.  A brief but helpful discussion of Word2Vec can be found at  https://spark.apache.org/docs/2.3.0/mllib-feature-extraction.html#word2vec, and many online resources are available.

Training should take around 10 minutes.

Defaults
- min count = 5
    - The minimum number of times a word must appear in the corpus to be counted as vocabulary
- vector size = 100
    - The dimension of the output vector. 
- Window Size is 5
    - The number of words in a window used for input training.  

In [8]:
val t0TermCount = System.nanoTime
val word2Vec = new Word2Vec().
    setInputCol("text").
    setOutputCol("result").
    setVectorSize(100).
    setMinCount(10)
val model = word2Vec.fit(documentDF)
val dtTermCount = round(System.nanoTime - t0TermCount)
println("time to generate word2vec: " + dtTermCount + "seconds")

time to generate word2vec: 680.152seconds


t0TermCount = 3226194910410768
word2Vec = w2v_9b6a30b49a60
model = w2v_9b6a30b49a60
dtTermCount = 680.152


680.152

The application of the Word2Vec algorithm has the following syntax in the ML Pipeline API

In [9]:
val result = model.transform(documentDF)

result = [text: array<string>, result: vector]


[text: array<string>, result: vector]

In [10]:
result.printSchema

root
 |-- text: array (nullable = true)
 |    |-- element: string (containsNull = true)
 |-- result: vector (nullable = true)



The left column is the processed article and the right column is the vector representation of the collection of words.  While the input can have many words (using a "one hot encoding" style of input), the right column always has the same number of features (100 in this case).

In [11]:
result.show()

+--------------------+--------------------+
|                text|              result|
+--------------------+--------------------+
|[clinton, press, ...|[-0.0502035705813...|
|[engineer, cowboy...|[-0.0190240844676...|
|[wiretapping, ini...|[-0.0232437958475...|
|[tieing, abortion...|[-0.0046298521373...|
|[expansion, mcgil...|[0.04100256933167...|
|[welcome, police,...|[0.00462032424805...|
|[role, national, ...|[0.02762026362249...|
|[concentrate, chi...|[-0.0356977651354...|
|[welcome, police,...|[-0.0049290160980...|
|[holocaust, memor...|[-0.0448481236408...|
|[real, source, tr...|[-0.0652494278806...|
|[study, percentag...|[0.00148526637805...|
|[abortion, privat...|[-0.0511604974162...|
|[earth, also, cel...|[-0.0254515306928...|
|[limiting, govt, ...|[-0.0213143074118...|
|[study, percentag...|[0.00635822871866...|
|[good, neighbor, ...|[-0.0191137050073...|
|[proof, viability...|[-0.0847958023401...|
|[proof, viability...|[-8.9741921466257...|
|[race, violence, ...|[-0.005263

Here we show the first article only:

In [12]:
result.show(1,false)

+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Now that we have a representation of our words as feature vectors, we can cluster them into "topics", or clusters, and begin to explore the internal structure that the Word2Vec and KMeans algorithms reveal.

In [13]:
//val K = math.floor(math.sqrt(documentDF.count/2)).toInt
val K = 10

K = 10


10

The KMeans algorithm is called in this notebook using the ML Pipeline API, which leverages the DataFrames API and has better performance in general.  

In [14]:
import org.apache.spark.ml.clustering.KMeans

In [15]:
val t0ClusterCount = System.nanoTime
         
val resultW2V = result.withColumnRenamed("result","features")
val kmeans = new KMeans().setK(K).setSeed(1L)
val modelK = kmeans.fit(resultW2V)
val labelsDF = modelK.transform(resultW2V).withColumnRenamed("prediction","labels")


val dtClusterCount = round(System.nanoTime - t0ClusterCount)
println("number of Clusters (k) used: " + K)
println("time to generate cluster: " + dtTermCount + "seconds")

number of Clusters (k) used: 10
time to generate cluster: 680.152seconds


t0ClusterCount = 3226884268615392
resultW2V = [text: array<string>, features: vector]
kmeans = kmeans_e2aaf0f83e0d
modelK = kmeans_e2aaf0f83e0d
labelsDF = [text: array<string>, features: vector ... 1 more field]
dtClusterCount = 337.434


337.434

In [16]:
resultW2V.show()

+--------------------+--------------------+
|                text|            features|
+--------------------+--------------------+
|[clinton, press, ...|[-0.0502035705813...|
|[engineer, cowboy...|[-0.0190240844676...|
|[wiretapping, ini...|[-0.0232437958475...|
|[tieing, abortion...|[-0.0046298521373...|
|[expansion, mcgil...|[0.04100256933167...|
|[welcome, police,...|[0.00462032424805...|
|[role, national, ...|[0.02762026362249...|
|[concentrate, chi...|[-0.0356977651354...|
|[welcome, police,...|[-0.0049290160980...|
|[holocaust, memor...|[-0.0448481236408...|
|[real, source, tr...|[-0.0652494278806...|
|[study, percentag...|[0.00148526637805...|
|[abortion, privat...|[-0.0511604974162...|
|[earth, also, cel...|[-0.0254515306928...|
|[limiting, govt, ...|[-0.0213143074118...|
|[study, percentag...|[0.00635822871866...|
|[good, neighbor, ...|[-0.0191137050073...|
|[proof, viability...|[-0.0847958023401...|
|[proof, viability...|[-8.9741921466257...|
|[race, violence, ...|[-0.005263

We can now display each article, along with its feature vector, and the cluster number that it was assigned to.

In [17]:
labelsDF.show()

+--------------------+--------------------+------+
|                text|            features|labels|
+--------------------+--------------------+------+
|[clinton, press, ...|[-0.0502035705813...|     3|
|[engineer, cowboy...|[-0.0190240844676...|     2|
|[wiretapping, ini...|[-0.0232437958475...|     3|
|[tieing, abortion...|[-0.0046298521373...|     2|
|[expansion, mcgil...|[0.04100256933167...|     2|
|[welcome, police,...|[0.00462032424805...|     3|
|[role, national, ...|[0.02762026362249...|     2|
|[concentrate, chi...|[-0.0356977651354...|     2|
|[welcome, police,...|[-0.0049290160980...|     2|
|[holocaust, memor...|[-0.0448481236408...|     7|
|[real, source, tr...|[-0.0652494278806...|     7|
|[study, percentag...|[0.00148526637805...|     2|
|[abortion, privat...|[-0.0511604974162...|     3|
|[earth, also, cel...|[-0.0254515306928...|     7|
|[limiting, govt, ...|[-0.0213143074118...|     3|
|[study, percentag...|[0.00635822871866...|     2|
|[good, neighbor, ...|[-0.01911

Let's take a look at the first cluster:

In [18]:
labelsDF.filter("labels = 0").show()

+--------------------+--------------------+------+
|                text|            features|labels|
+--------------------+--------------------+------+
|[hitachi, oscilli...|[-0.0834846808010...|     0|
|[atari, processor...|[-0.0448761401124...|     0|
|[voice, synth, ch...|[-0.0995623438308...|     0|
|[oscilloscope, sa...|[-0.1196322398347...|     0|
|[circuit, inquiri...|[-0.0765004193420...|     0|
|[dram, controller...|[-0.0587306980641...|     0|
|[need, help, find...|[-0.0456011835331...|     0|
|[oscilliscopes, s...|[-0.1171103589321...|     0|
|[casio, serial, l...|[-0.0594486877150...|     0|
|[cross, compilers...|[-0.0312366772975...|     0|
|[video, free, uni...|[-0.0092522239512...|     0|
|[circuit, boards,...|[-0.0191166970813...|     0|
|[drive, array, wo...|[-0.0032491331498...|     0|
|[remote, extender...|[-0.0429821987527...|     0|
|[amplifier, into,...|[-0.0051534950528...|     0|
|[power, supply, s...|[0.00759276533217...|     0|
|[telephone, circu...|[-0.03784

That view only shows us the first few words, so let's use the expanded display version:

In [19]:
labelsDF.select("text").filter("labels = 0").show(false)

+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

It sure looks like this particular cluster maps very closely to the contents of the `sci.electronics` directory.  Perhaps this cluster could be a similar topic?  It seems like a reasonable place to start when trying to work through the text. 

Let's take a look at how many articles are in each topic:

In [21]:
labelsDF.groupBy("labels").count().orderBy("labels").show()

+------+-----+
|labels|count|
+------+-----+
|     0| 1278|
|     1| 1222|
|     2| 3568|
|     3| 2330|
|     4|  481|
|     5| 1521|
|     6| 3904|
|     7| 2180|
|     8| 2888|
|     9|  625|
+------+-----+



One very nice result here is that the clusters seem to be somewhat uniformly distributed...each cluster has a significant population, and no one cluster seems to have aggregated too many data points.  It's only the beginning here, but we might be able to identify topics with this approach!

In [22]:
labelsDF.count()

19997

### Explore Work Similarities

In [23]:
val wordVectorsDF = model.getVectors
val vocabSize = wordVectorsDF.count()
println("Vocabulary Size: "+vocabSize)

Vocabulary Size: 14896


wordVectorsDF = [word: string, vector: vector]
vocabSize = 14896


14896

In [24]:
wordVectorsDF.printSchema()

root
 |-- word: string (nullable = true)
 |-- vector: vector (nullable = true)



In [25]:
wordVectorsDF.select("word").show()

+------------+
|        word|
+------------+
|   professed|
|    incident|
|     misfire|
|     serious|
|     acronym|
|      comply|
|      breaks|
|   forgotten|
|    precious|
|       mario|
|   controler|
|inflammatory|
|     sectors|
|   ascension|
|      teresa|
|       boggs|
|       tripe|
|    embedded|
|  respecting|
|      daigle|
+------------+
only showing top 20 rows



In [26]:
model.explainParams()

vectorSize: the dimension of codes after transforming from wo...


inputCol: input column name (current: text)
maxIter: maximum number of iterations (>= 0) (default: 1)
maxSentenceLength: Maximum length (in words) of each sentence in the input data. Any sentence longer than this threshold will be divided into chunks up to the size (> 0) (default: 1000)
minCount: the minimum number of times a token must appear to be included in the word2vec model's vocabulary (>= 0) (default: 5, current: 10)
numPartitions: number of partitions for sentences of words (> 0) (default: 1)
outputCol: output column name (default: w2v_9b6a30b49a60__output, current: result)
seed: random seed (default: -1961189076)
stepSize: Step size to be used for each iteration of optimization (> 0) (default: 0.025)
vectorSize: the dimension of codes after transforming from words (> 0) (default: 100, current: 100)
windowSize: the window size (context words from [-window, window]) (> 0) (default: 5)

Word2Vec is an interesting model in that it can take an arbitrary number of words and return a vector.  The input is one very large vector of $1$s and $0$s, where each row of the vector corresponts to a word in the learned vocabulary.  Thus, an article can have many words that go into the input vector, or it can have only a single word.  The single word inputs are interesting because they can be used to find synonyms.  Let's find some synonyms to some words from our learned model:

In [27]:
val word = "drive"
val topN = 5
val synonymsDF = model.findSynonyms(word, topN)
synonymsDF.show()

+----------+------------------+
|      word|        similarity|
+----------+------------------+
|    floppy|0.7424999475479126|
|    drives|0.7419173121452332|
|controller|0.7130770087242126|
|      scsi|0.7129872441291809|
|      esdi|0.6894485950469971|
+----------+------------------+



word = drive
topN = 5
synonymsDF = [word: string, similarity: double]


[word: string, similarity: double]

In [28]:
val word = "network"
val topN = 5
val synonymsDF = model.findSynonyms(word, topN)
synonymsDF.show()

+---------+------------------+
|     word|        similarity|
+---------+------------------+
|     arpa|0.6670048832893372|
|workplace|  0.61444491147995|
|  pathway|0.5837959051132202|
|   novell|0.5518525838851929|
|   encore|0.5424853563308716|
+---------+------------------+



word = network
topN = 5
synonymsDF = [word: string, similarity: double]


[word: string, similarity: double]

Finally, let's save our work and look at another topic modeling approach...the Latent Dirichlet Allocation!

In [29]:
model.write.overwrite.save("model/word2vec")